// @angular/compiler-cli/src/ngtsc/file_system/src/invalid_file_system.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/file_system/src/invalid_file_system", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.InvalidFileSystem = void 0;
    /**
     * The default `FileSystem` that will always fail.
     *
     * This is a way of ensuring that the developer consciously chooses and
     * configures the `FileSystem` before using it; particularly important when
     * considering static functions like `absoluteFrom()` which rely on
     * the `FileSystem` under the hood.
     */
    var InvalidFileSystem = /** @class */ (function () {
        function InvalidFileSystem() {
        }
        InvalidFileSystem.prototype.exists = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.readFile = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.readFileBuffer = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.writeFile = function (path, data, exclusive) {
            throw makeError();
        };
        InvalidFileSystem.prototype.removeFile = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.symlink = function (target, path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.readdir = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.lstat = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.stat = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.pwd = function () {
            throw makeError();
        };
        InvalidFileSystem.prototype.chdir = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.extname = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.copyFile = function (from, to) {
            throw makeError();
        };
        InvalidFileSystem.prototype.moveFile = function (from, to) {
            throw makeError();
        };
        InvalidFileSystem.prototype.ensureDir = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.removeDeep = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.isCaseSensitive = function () {
            throw makeError();
        };
        InvalidFileSystem.prototype.resolve = function () {
            var paths = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                paths[_i] = arguments[_i];
            }
            throw makeError();
        };
        InvalidFileSystem.prototype.dirname = function (file) {
            throw makeError();
        };
        InvalidFileSystem.prototype.join = function (basePath) {
            var paths = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                paths[_i - 1] = arguments[_i];
            }
            throw makeError();
        };
        InvalidFileSystem.prototype.isRoot = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.isRooted = function (path) {
            throw makeError();
        };
        InvalidFileSystem.prototype.relative = function (from, to) {
            throw makeError();
        };
        InvalidFileSystem.prototype.basename = function (filePath, extension) {
            throw makeError();
        };
        InvalidFileSystem.prototype.realpath = function (filePath) {
            throw makeError();
        };
        InvalidFileSystem.prototype.getDefaultLibLocation = function () {
            throw makeError();
        };
        InvalidFileSystem.prototype.normalize = function (path) {
            throw makeError();
        };
        return InvalidFileSystem;
    }());
    exports.InvalidFileSystem = InvalidFileSystem;
    function makeError() {
        return new Error('FileSystem has not been configured. Please call `setFileSystem()` before calling this method.');
    }
});
