// @angular/cdk/schematics/update-tool/target-version.mjs
"use strict";
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllVersionNames = exports.TargetVersion = void 0;
/** Possible versions that can be automatically migrated by `ng update`. */
// Used in an `Object.keys` call below so it can't be `const enum`.
// tslint:disable-next-line:prefer-const-enum
var TargetVersion;
(function (TargetVersion) {
    TargetVersion["V6"] = "version 6";
    TargetVersion["V7"] = "version 7";
    TargetVersion["V8"] = "version 8";
    TargetVersion["V9"] = "version 9";
    TargetVersion["V10"] = "version 10";
    TargetVersion["V11"] = "version 11";
    TargetVersion["V12"] = "version 12";
})(TargetVersion = exports.TargetVersion || (exports.TargetVersion = {}));
/**
 * Returns all versions that are supported by "ng update". The versions are determined
 * based on the "TargetVersion" enum.
 */
function getAllVersionNames() {
    return Object.keys(TargetVersion).filter(enumValue => {
        return typeof TargetVersion[enumValue] === 'string';
    });
}
exports.getAllVersionNames = getAllVersionNames;
