// @angular/compiler-cli/linker/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker", ["require", "exports", "@angular/compiler-cli/linker/src/ast/utils", "@angular/compiler-cli/linker/src/fatal_linker_error", "@angular/compiler-cli/linker/src/file_linker/file_linker", "@angular/compiler-cli/linker/src/file_linker/linker_environment", "@angular/compiler-cli/linker/src/file_linker/linker_options", "@angular/compiler-cli/linker/src/file_linker/needs_linking"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.needsLinking = exports.DEFAULT_LINKER_OPTIONS = exports.LinkerEnvironment = exports.FileLinker = exports.isFatalLinkerError = exports.FatalLinkerError = exports.assert = void 0;
    var utils_1 = require("@angular/compiler-cli/linker/src/ast/utils");
    Object.defineProperty(exports, "assert", { enumerable: true, get: function () { return utils_1.assert; } });
    var fatal_linker_error_1 = require("@angular/compiler-cli/linker/src/fatal_linker_error");
    Object.defineProperty(exports, "FatalLinkerError", { enumerable: true, get: function () { return fatal_linker_error_1.FatalLinkerError; } });
    Object.defineProperty(exports, "isFatalLinkerError", { enumerable: true, get: function () { return fatal_linker_error_1.isFatalLinkerError; } });
    var file_linker_1 = require("@angular/compiler-cli/linker/src/file_linker/file_linker");
    Object.defineProperty(exports, "FileLinker", { enumerable: true, get: function () { return file_linker_1.FileLinker; } });
    var linker_environment_1 = require("@angular/compiler-cli/linker/src/file_linker/linker_environment");
    Object.defineProperty(exports, "LinkerEnvironment", { enumerable: true, get: function () { return linker_environment_1.LinkerEnvironment; } });
    var linker_options_1 = require("@angular/compiler-cli/linker/src/file_linker/linker_options");
    Object.defineProperty(exports, "DEFAULT_LINKER_OPTIONS", { enumerable: true, get: function () { return linker_options_1.DEFAULT_LINKER_OPTIONS; } });
    var needs_linking_1 = require("@angular/compiler-cli/linker/src/file_linker/needs_linking");
    Object.defineProperty(exports, "needsLinking", { enumerable: true, get: function () { return needs_linking_1.needsLinking; } });
});
