// @angular/material/esm2015/datepicker/testing/date-range-input-harness.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
import { HarnessPredicate, parallel, TestKey } from '@angular/cdk/testing';
import { MatDatepickerInputHarnessBase, getInputPredicate } from './datepicker-input-harness-base';
import { DatepickerTriggerHarnessBase } from './datepicker-trigger-harness-base';
/** Harness for interacting with a standard Material date range start input in tests. */
export class MatStartDateHarness extends MatDatepickerInputHarnessBase {
    /**
     * Gets a `HarnessPredicate` that can be used to search for a `MatStartDateHarness`
     * that meets certain criteria.
     * @param options Options for filtering which input instances are considered a match.
     * @return a `HarnessPredicate` configured with the given options.
     */
    static with(options = {}) {
        return getInputPredicate(MatStartDateHarness, options);
    }
}
MatStartDateHarness.hostSelector = '.mat-start-date';
/** Harness for interacting with a standard Material date range end input in tests. */
export class MatEndDateHarness extends MatDatepickerInputHarnessBase {
    /**
     * Gets a `HarnessPredicate` that can be used to search for a `MatEndDateHarness`
     * that meets certain criteria.
     * @param options Options for filtering which input instances are considered a match.
     * @return a `HarnessPredicate` configured with the given options.
     */
    static with(options = {}) {
        return getInputPredicate(MatEndDateHarness, options);
    }
}
MatEndDateHarness.hostSelector = '.mat-end-date';
/** Harness for interacting with a standard Material date range input in tests. */
export class MatDateRangeInputHarness extends DatepickerTriggerHarnessBase {
    /**
     * Gets a `HarnessPredicate` that can be used to search for a `MatDateRangeInputHarness`
     * that meets certain criteria.
     * @param options Options for filtering which input instances are considered a match.
     * @return a `HarnessPredicate` configured with the given options.
     */
    static with(options = {}) {
        return new HarnessPredicate(MatDateRangeInputHarness, options)
            .addOption('value', options.value, (harness, value) => HarnessPredicate.stringMatches(harness.getValue(), value));
    }
    /** Gets the combined value of the start and end inputs, including the separator. */
    getValue() {
        return __awaiter(this, void 0, void 0, function* () {
            const [start, end, separator] = yield parallel(() => [
                this.getStartInput().then(input => input.getValue()),
                this.getEndInput().then(input => input.getValue()),
                this.getSeparator()
            ]);
            return start + `${end ? ` ${separator} ${end}` : ''}`;
        });
    }
    /** Gets the inner start date input inside the range input. */
    getStartInput() {
        return __awaiter(this, void 0, void 0, function* () {
            // Don't pass in filters here since the start input is required and there can only be one.
            return this.locatorFor(MatStartDateHarness)();
        });
    }
    /** Gets the inner start date input inside the range input. */
    getEndInput() {
        return __awaiter(this, void 0, void 0, function* () {
            // Don't pass in filters here since the end input is required and there can only be one.
            return this.locatorFor(MatEndDateHarness)();
        });
    }
    /** Gets the separator text between the values of the two inputs. */
    getSeparator() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.locatorFor('.mat-date-range-input-separator')()).text();
        });
    }
    /** Gets whether the range input is disabled. */
    isDisabled() {
        return __awaiter(this, void 0, void 0, function* () {
            // We consider the input as disabled if both of the sub-inputs are disabled.
            const [startDisabled, endDisabled] = yield parallel(() => [
                this.getStartInput().then(input => input.isDisabled()),
                this.getEndInput().then(input => input.isDisabled())
            ]);
            return startDisabled && endDisabled;
        });
    }
    /** Gets whether the range input is required. */
    isRequired() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.host()).hasClass('mat-date-range-input-required');
        });
    }
    /** Opens the calendar associated with the input. */
    isCalendarOpen() {
        return __awaiter(this, void 0, void 0, function* () {
            // `aria-owns` is set on both inputs only if there's an
            // open range picker so we can use it as an indicator.
            const startHost = yield (yield this.getStartInput()).host();
            return (yield startHost.getAttribute('aria-owns')) != null;
        });
    }
    _openCalendar() {
        return __awaiter(this, void 0, void 0, function* () {
            // Alt + down arrow is the combination for opening the calendar with the keyboard.
            const startHost = yield (yield this.getStartInput()).host();
            return startHost.sendKeys({ alt: true }, TestKey.DOWN_ARROW);
        });
    }
}
MatDateRangeInputHarness.hostSelector = '.mat-date-range-input';
