// @angular/material/esm2015/datepicker/testing/datepicker-trigger-harness-base.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
import { ComponentHarness, parallel } from '@angular/cdk/testing';
import { MatCalendarHarness } from './calendar-harness';
/** Base class for harnesses that can trigger a calendar. */
export class DatepickerTriggerHarnessBase extends ComponentHarness {
    /** Opens the calendar if the trigger is enabled and it has a calendar. */
    openCalendar() {
        return __awaiter(this, void 0, void 0, function* () {
            const [isDisabled, hasCalendar] = yield parallel(() => [this.isDisabled(), this.hasCalendar()]);
            if (!isDisabled && hasCalendar) {
                return this._openCalendar();
            }
        });
    }
    /** Closes the calendar if it is open. */
    closeCalendar() {
        return __awaiter(this, void 0, void 0, function* () {
            if (yield this.isCalendarOpen()) {
                yield closeCalendar(getCalendarId(this.host()), this.documentRootLocatorFactory());
                // This is necessary so that we wait for the closing animation to finish in touch UI mode.
                yield this.forceStabilize();
            }
        });
    }
    /** Gets whether there is a calendar associated with the trigger. */
    hasCalendar() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield getCalendarId(this.host())) != null;
        });
    }
    /**
     * Gets the `MatCalendarHarness` that is associated with the trigger.
     * @param filter Optionally filters which calendar is included.
     */
    getCalendar(filter = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return getCalendar(filter, this.host(), this.documentRootLocatorFactory());
        });
    }
}
/** Gets the ID of the calendar that a particular test element can trigger. */
export function getCalendarId(host) {
    return __awaiter(this, void 0, void 0, function* () {
        return (yield host).getAttribute('data-mat-calendar');
    });
}
/** Closes the calendar with a specific ID. */
export function closeCalendar(calendarId, documentLocator) {
    return __awaiter(this, void 0, void 0, function* () {
        // We close the calendar by clicking on the backdrop, even though all datepicker variants
        // have the ability to close by pressing escape. The backdrop is preferrable, because the
        // escape key has multiple functions inside a range picker (either cancel the current range
        // or close the calendar). Since we don't have access to set the ID on the backdrop in all
        // cases, we set a unique class instead which is the same as the calendar's ID and suffixed
        // with `-backdrop`.
        const backdropSelector = `.${yield calendarId}-backdrop`;
        return (yield documentLocator.locatorFor(backdropSelector)()).click();
    });
}
/** Gets the test harness for a calendar associated with a particular host. */
export function getCalendar(filter, host, documentLocator) {
    return __awaiter(this, void 0, void 0, function* () {
        const calendarId = yield getCalendarId(host);
        if (!calendarId) {
            throw Error(`Element is not associated with a calendar`);
        }
        return documentLocator.locatorFor(MatCalendarHarness.with(Object.assign(Object.assign({}, filter), { selector: `#${calendarId}` })))();
    });
}
