// @angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_native.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_native", ["require", "exports", "tslib", "os", "@angular/compiler-cli/src/ngtsc/file_system/src/node_js_file_system", "@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.MockFileSystemNative = void 0;
    var tslib_1 = require("tslib");
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    /// <reference types="node" />
    var os = require("os");
    var node_js_file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/node_js_file_system");
    var mock_file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system");
    var isWindows = os.platform() === 'win32';
    var MockFileSystemNative = /** @class */ (function (_super) {
        tslib_1.__extends(MockFileSystemNative, _super);
        function MockFileSystemNative(cwd) {
            if (cwd === void 0) { cwd = '/'; }
            return _super.call(this, undefined, cwd) || this;
        }
        // Delegate to the real NodeJSFileSystem for these path related methods
        MockFileSystemNative.prototype.resolve = function () {
            var _a;
            var paths = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                paths[_i] = arguments[_i];
            }
            return (_a = node_js_file_system_1.NodeJSFileSystem.prototype.resolve).call.apply(_a, tslib_1.__spreadArray([this, this.pwd()], tslib_1.__read(paths)));
        };
        MockFileSystemNative.prototype.dirname = function (file) {
            return node_js_file_system_1.NodeJSFileSystem.prototype.dirname.call(this, file);
        };
        MockFileSystemNative.prototype.join = function (basePath) {
            var _a;
            var paths = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                paths[_i - 1] = arguments[_i];
            }
            return (_a = node_js_file_system_1.NodeJSFileSystem.prototype.join).call.apply(_a, tslib_1.__spreadArray([this, basePath], tslib_1.__read(paths)));
        };
        MockFileSystemNative.prototype.relative = function (from, to) {
            return node_js_file_system_1.NodeJSFileSystem.prototype.relative.call(this, from, to);
        };
        MockFileSystemNative.prototype.basename = function (filePath, extension) {
            return node_js_file_system_1.NodeJSFileSystem.prototype.basename.call(this, filePath, extension);
        };
        MockFileSystemNative.prototype.isCaseSensitive = function () {
            return node_js_file_system_1.NodeJSFileSystem.prototype.isCaseSensitive.call(this);
        };
        MockFileSystemNative.prototype.isRooted = function (path) {
            return node_js_file_system_1.NodeJSFileSystem.prototype.isRooted.call(this, path);
        };
        MockFileSystemNative.prototype.isRoot = function (path) {
            return node_js_file_system_1.NodeJSFileSystem.prototype.isRoot.call(this, path);
        };
        MockFileSystemNative.prototype.normalize = function (path) {
            // When running in Windows, absolute paths are normalized to always include a drive letter. This
            // ensures that rooted posix paths used in tests will be normalized to real Windows paths, i.e.
            // including a drive letter. Note that the same normalization is done in emulated Windows mode
            // (see `MockFileSystemWindows`) so that the behavior is identical between native Windows and
            // emulated Windows mode.
            if (isWindows) {
                path = path.replace(/^[\/\\]/i, 'C:/');
            }
            return node_js_file_system_1.NodeJSFileSystem.prototype.normalize.call(this, path);
        };
        MockFileSystemNative.prototype.splitPath = function (path) {
            return path.split(/[\\\/]/);
        };
        return MockFileSystemNative;
    }(mock_file_system_1.MockFileSystem));
    exports.MockFileSystemNative = MockFileSystemNative;
});
