// @angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system", ["require", "exports", "tslib", "@angular/compiler-cli/src/ngtsc/file_system/src/helpers"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.isFolder = exports.isSymLink = exports.isFile = exports.SymLink = exports.MockFileSystem = void 0;
    var tslib_1 = require("tslib");
    var helpers_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/helpers");
    /**
     * An in-memory file system that can be used in unit tests.
     */
    var MockFileSystem = /** @class */ (function () {
        function MockFileSystem(_isCaseSensitive, cwd) {
            if (_isCaseSensitive === void 0) { _isCaseSensitive = false; }
            if (cwd === void 0) { cwd = '/'; }
            this._isCaseSensitive = _isCaseSensitive;
            this._fileTree = {};
            this._cwd = this.normalize(cwd);
        }
        MockFileSystem.prototype.isCaseSensitive = function () {
            return this._isCaseSensitive;
        };
        MockFileSystem.prototype.exists = function (path) {
            return this.findFromPath(path).entity !== null;
        };
        MockFileSystem.prototype.readFile = function (path) {
            var entity = this.findFromPath(path).entity;
            if (isFile(entity)) {
                return entity.toString();
            }
            else {
                throw new MockFileSystemError('ENOENT', path, "File \"" + path + "\" does not exist.");
            }
        };
        MockFileSystem.prototype.readFileBuffer = function (path) {
            var entity = this.findFromPath(path).entity;
            if (isFile(entity)) {
                return entity instanceof Uint8Array ? entity : new Buffer(entity);
            }
            else {
                throw new MockFileSystemError('ENOENT', path, "File \"" + path + "\" does not exist.");
            }
        };
        MockFileSystem.prototype.writeFile = function (path, data, exclusive) {
            if (exclusive === void 0) { exclusive = false; }
            var _a = tslib_1.__read(this.splitIntoFolderAndFile(path), 2), folderPath = _a[0], basename = _a[1];
            var entity = this.findFromPath(folderPath).entity;
            if (entity === null || !isFolder(entity)) {
                throw new MockFileSystemError('ENOENT', path, "Unable to write file \"" + path + "\". The containing folder does not exist.");
            }
            if (exclusive && entity[basename] !== undefined) {
                throw new MockFileSystemError('EEXIST', path, "Unable to exclusively write file \"" + path + "\". The file already exists.");
            }
            entity[basename] = data;
        };
        MockFileSystem.prototype.removeFile = function (path) {
            var _a = tslib_1.__read(this.splitIntoFolderAndFile(path), 2), folderPath = _a[0], basename = _a[1];
            var entity = this.findFromPath(folderPath).entity;
            if (entity === null || !isFolder(entity)) {
                throw new MockFileSystemError('ENOENT', path, "Unable to remove file \"" + path + "\". The containing folder does not exist.");
            }
            if (isFolder(entity[basename])) {
                throw new MockFileSystemError('EISDIR', path, "Unable to remove file \"" + path + "\". The path to remove is a folder.");
            }
            delete entity[basename];
        };
        MockFileSystem.prototype.symlink = function (target, path) {
            var _a = tslib_1.__read(this.splitIntoFolderAndFile(path), 2), folderPath = _a[0], basename = _a[1];
            var entity = this.findFromPath(folderPath).entity;
            if (entity === null || !isFolder(entity)) {
                throw new MockFileSystemError('ENOENT', path, "Unable to create symlink at \"" + path + "\". The containing folder does not exist.");
            }
            entity[basename] = new SymLink(target);
        };
        MockFileSystem.prototype.readdir = function (path) {
            var entity = this.findFromPath(path).entity;
            if (entity === null) {
                throw new MockFileSystemError('ENOENT', path, "Unable to read directory \"" + path + "\". It does not exist.");
            }
            if (isFile(entity)) {
                throw new MockFileSystemError('ENOTDIR', path, "Unable to read directory \"" + path + "\". It is a file.");
            }
            return Object.keys(entity);
        };
        MockFileSystem.prototype.lstat = function (path) {
            var entity = this.findFromPath(path).entity;
            if (entity === null) {
                throw new MockFileSystemError('ENOENT', path, "File \"" + path + "\" does not exist.");
            }
            return new MockFileStats(entity);
        };
        MockFileSystem.prototype.stat = function (path) {
            var entity = this.findFromPath(path, { followSymLinks: true }).entity;
            if (entity === null) {
                throw new MockFileSystemError('ENOENT', path, "File \"" + path + "\" does not exist.");
            }
            return new MockFileStats(entity);
        };
        MockFileSystem.prototype.copyFile = function (from, to) {
            this.writeFile(to, this.readFile(from));
        };
        MockFileSystem.prototype.moveFile = function (from, to) {
            this.writeFile(to, this.readFile(from));
            var result = this.findFromPath(helpers_1.dirname(from));
            var folder = result.entity;
            var name = helpers_1.basename(from);
            delete folder[name];
        };
        MockFileSystem.prototype.ensureDir = function (path) {
            var e_1, _a;
            var _this = this;
            var segments = this.splitPath(path).map(function (segment) { return _this.getCanonicalPath(segment); });
            // Convert the root folder to a canonical empty string `''` (on Windows it would be `'C:'`).
            segments[0] = '';
            if (segments.length > 1 && segments[segments.length - 1] === '') {
                // Remove a trailing slash (unless the path was only `/`)
                segments.pop();
            }
            var current = this._fileTree;
            try {
                for (var segments_1 = tslib_1.__values(segments), segments_1_1 = segments_1.next(); !segments_1_1.done; segments_1_1 = segments_1.next()) {
                    var segment = segments_1_1.value;
                    if (isFile(current[segment])) {
                        throw new Error("Folder already exists as a file.");
                    }
                    if (!current[segment]) {
                        current[segment] = {};
                    }
                    current = current[segment];
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (segments_1_1 && !segments_1_1.done && (_a = segments_1.return)) _a.call(segments_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return current;
        };
        MockFileSystem.prototype.removeDeep = function (path) {
            var _a = tslib_1.__read(this.splitIntoFolderAndFile(path), 2), folderPath = _a[0], basename = _a[1];
            var entity = this.findFromPath(folderPath).entity;
            if (entity === null || !isFolder(entity)) {
                throw new MockFileSystemError('ENOENT', path, "Unable to remove folder \"" + path + "\". The containing folder does not exist.");
            }
            delete entity[basename];
        };
        MockFileSystem.prototype.isRoot = function (path) {
            return this.dirname(path) === path;
        };
        MockFileSystem.prototype.extname = function (path) {
            var match = /.+(\.[^.]*)$/.exec(path);
            return match !== null ? match[1] : '';
        };
        MockFileSystem.prototype.realpath = function (filePath) {
            var result = this.findFromPath(filePath, { followSymLinks: true });
            if (result.entity === null) {
                throw new MockFileSystemError('ENOENT', filePath, "Unable to find the real path of \"" + filePath + "\". It does not exist.");
            }
            else {
                return result.path;
            }
        };
        MockFileSystem.prototype.pwd = function () {
            return this._cwd;
        };
        MockFileSystem.prototype.chdir = function (path) {
            this._cwd = this.normalize(path);
        };
        MockFileSystem.prototype.getDefaultLibLocation = function () {
            // Mimic the node module resolution algorithm and start in the current directory, then look
            // progressively further up the tree until reaching the FS root.
            // E.g. if the current directory is /foo/bar, look in /foo/bar/node_modules, then
            // /foo/node_modules, then /node_modules.
            var path = 'node_modules/typescript/lib';
            var resolvedPath = this.resolve(path);
            // Construct a path for the top-level node_modules to identify the stopping point.
            var topLevelNodeModules = this.resolve('/' + path);
            while (resolvedPath !== topLevelNodeModules) {
                if (this.exists(resolvedPath)) {
                    return resolvedPath;
                }
                // Not here, look one level higher.
                path = '../' + path;
                resolvedPath = this.resolve(path);
            }
            // The loop exits before checking the existence of /node_modules/typescript at the top level.
            // This is intentional - if no /node_modules/typescript exists anywhere in the tree, there's
            // nothing this function can do about it, and TS may error later if it looks for a lib.d.ts file
            // within this directory. It might be okay, though, if TS never checks for one.
            return topLevelNodeModules;
        };
        MockFileSystem.prototype.dump = function () {
            var entity = this.findFromPath(this.resolve('/')).entity;
            if (entity === null || !isFolder(entity)) {
                return {};
            }
            return this.cloneFolder(entity);
        };
        MockFileSystem.prototype.init = function (folder) {
            this.mount(this.resolve('/'), folder);
        };
        MockFileSystem.prototype.mount = function (path, folder) {
            if (this.exists(path)) {
                throw new Error("Unable to mount in '" + path + "' as it already exists.");
            }
            var mountFolder = this.ensureDir(path);
            this.copyInto(folder, mountFolder);
        };
        MockFileSystem.prototype.cloneFolder = function (folder) {
            var clone = {};
            this.copyInto(folder, clone);
            return clone;
        };
        MockFileSystem.prototype.copyInto = function (from, to) {
            for (var path in from) {
                var item = from[path];
                var canonicalPath = this.getCanonicalPath(path);
                if (isSymLink(item)) {
                    to[canonicalPath] = new SymLink(this.getCanonicalPath(item.path));
                }
                else if (isFolder(item)) {
                    to[canonicalPath] = this.cloneFolder(item);
                }
                else {
                    to[canonicalPath] = from[path];
                }
            }
        };
        MockFileSystem.prototype.findFromPath = function (path, options) {
            var followSymLinks = !!options && options.followSymLinks;
            var segments = this.splitPath(path);
            if (segments.length > 1 && segments[segments.length - 1] === '') {
                // Remove a trailing slash (unless the path was only `/`)
                segments.pop();
            }
            // Convert the root folder to a canonical empty string `""` (on Windows it would be `C:`).
            segments[0] = '';
            var current = this._fileTree;
            while (segments.length) {
                current = current[this.getCanonicalPath(segments.shift())];
                if (current === undefined) {
                    return { path: path, entity: null };
                }
                if (segments.length > 0 && (!isFolder(current))) {
                    current = null;
                    break;
                }
                if (isFile(current)) {
                    break;
                }
                if (isSymLink(current)) {
                    if (followSymLinks) {
                        return this.findFromPath(helpers_1.resolve.apply(void 0, tslib_1.__spreadArray([current.path], tslib_1.__read(segments))), { followSymLinks: followSymLinks });
                    }
                    else {
                        break;
                    }
                }
            }
            return { path: path, entity: current };
        };
        MockFileSystem.prototype.splitIntoFolderAndFile = function (path) {
            var segments = this.splitPath(this.getCanonicalPath(path));
            var file = segments.pop();
            return [path.substring(0, path.length - file.length - 1), file];
        };
        MockFileSystem.prototype.getCanonicalPath = function (p) {
            return this.isCaseSensitive() ? p : p.toLowerCase();
        };
        return MockFileSystem;
    }());
    exports.MockFileSystem = MockFileSystem;
    var SymLink = /** @class */ (function () {
        function SymLink(path) {
            this.path = path;
        }
        return SymLink;
    }());
    exports.SymLink = SymLink;
    var MockFileStats = /** @class */ (function () {
        function MockFileStats(entity) {
            this.entity = entity;
        }
        MockFileStats.prototype.isFile = function () {
            return isFile(this.entity);
        };
        MockFileStats.prototype.isDirectory = function () {
            return isFolder(this.entity);
        };
        MockFileStats.prototype.isSymbolicLink = function () {
            return isSymLink(this.entity);
        };
        return MockFileStats;
    }());
    var MockFileSystemError = /** @class */ (function (_super) {
        tslib_1.__extends(MockFileSystemError, _super);
        function MockFileSystemError(code, path, message) {
            var _this = _super.call(this, message) || this;
            _this.code = code;
            _this.path = path;
            return _this;
        }
        return MockFileSystemError;
    }(Error));
    function isFile(item) {
        return Buffer.isBuffer(item) || typeof item === 'string';
    }
    exports.isFile = isFile;
    function isSymLink(item) {
        return item instanceof SymLink;
    }
    exports.isSymLink = isSymLink;
    function isFolder(item) {
        return item !== null && !isFile(item) && !isSymLink(item);
    }
    exports.isFolder = isFolder;
});
