// @angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_posix.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_posix", ["require", "exports", "tslib", "path", "@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.MockFileSystemPosix = void 0;
    var tslib_1 = require("tslib");
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    /// <reference types="node" />
    var p = require("path");
    var mock_file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system");
    var MockFileSystemPosix = /** @class */ (function (_super) {
        tslib_1.__extends(MockFileSystemPosix, _super);
        function MockFileSystemPosix() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        MockFileSystemPosix.prototype.resolve = function () {
            var _a;
            var paths = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                paths[_i] = arguments[_i];
            }
            var resolved = (_a = p.posix).resolve.apply(_a, tslib_1.__spreadArray([this.pwd()], tslib_1.__read(paths)));
            return this.normalize(resolved);
        };
        MockFileSystemPosix.prototype.dirname = function (file) {
            return this.normalize(p.posix.dirname(file));
        };
        MockFileSystemPosix.prototype.join = function (basePath) {
            var _a;
            var paths = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                paths[_i - 1] = arguments[_i];
            }
            return this.normalize((_a = p.posix).join.apply(_a, tslib_1.__spreadArray([basePath], tslib_1.__read(paths))));
        };
        MockFileSystemPosix.prototype.relative = function (from, to) {
            return this.normalize(p.posix.relative(from, to));
        };
        MockFileSystemPosix.prototype.basename = function (filePath, extension) {
            return p.posix.basename(filePath, extension);
        };
        MockFileSystemPosix.prototype.isRooted = function (path) {
            return path.startsWith('/');
        };
        MockFileSystemPosix.prototype.splitPath = function (path) {
            return path.split('/');
        };
        MockFileSystemPosix.prototype.normalize = function (path) {
            return path.replace(/^[a-z]:\//i, '/').replace(/\\/g, '/');
        };
        return MockFileSystemPosix;
    }(mock_file_system_1.MockFileSystem));
    exports.MockFileSystemPosix = MockFileSystemPosix;
});
