// @angular/compiler-cli/src/ngtsc/file_system/testing/src/test_helper.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/file_system/testing/src/test_helper", ["require", "exports", "tslib", "typescript", "@angular/compiler-cli/src/ngtsc/file_system/src/helpers", "@angular/compiler-cli/src/ngtsc/file_system/src/invalid_file_system", "@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_native", "@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_posix", "@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_windows"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.initMockFileSystem = exports.runInEachFileSystem = void 0;
    var tslib_1 = require("tslib");
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    /// <reference types="jasmine"/>
    var ts = require("typescript");
    var helpers_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/helpers");
    var invalid_file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/invalid_file_system");
    var mock_file_system_native_1 = require("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_native");
    var mock_file_system_posix_1 = require("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_posix");
    var mock_file_system_windows_1 = require("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_windows");
    var FS_NATIVE = 'Native';
    var FS_OS_X = 'OS/X';
    var FS_UNIX = 'Unix';
    var FS_WINDOWS = 'Windows';
    var FS_ALL = [FS_OS_X, FS_WINDOWS, FS_UNIX, FS_NATIVE];
    function runInEachFileSystemFn(callback) {
        FS_ALL.forEach(function (os) { return runInFileSystem(os, callback, false); });
    }
    function runInFileSystem(os, callback, error) {
        describe("<<FileSystem: " + os + ">>", function () {
            beforeEach(function () { return initMockFileSystem(os); });
            afterEach(function () { return helpers_1.setFileSystem(new invalid_file_system_1.InvalidFileSystem()); });
            callback(os);
            if (error) {
                afterAll(function () {
                    throw new Error("runInFileSystem limited to " + os + ", cannot pass");
                });
            }
        });
    }
    exports.runInEachFileSystem = runInEachFileSystemFn;
    exports.runInEachFileSystem.native = function (callback) {
        return runInFileSystem(FS_NATIVE, callback, true);
    };
    exports.runInEachFileSystem.osX = function (callback) {
        return runInFileSystem(FS_OS_X, callback, true);
    };
    exports.runInEachFileSystem.unix = function (callback) {
        return runInFileSystem(FS_UNIX, callback, true);
    };
    exports.runInEachFileSystem.windows = function (callback) {
        return runInFileSystem(FS_WINDOWS, callback, true);
    };
    function initMockFileSystem(os, cwd) {
        var fs = createMockFileSystem(os, cwd);
        helpers_1.setFileSystem(fs);
        monkeyPatchTypeScript(os, fs);
        return fs;
    }
    exports.initMockFileSystem = initMockFileSystem;
    function createMockFileSystem(os, cwd) {
        switch (os) {
            case 'OS/X':
                return new mock_file_system_posix_1.MockFileSystemPosix(/* isCaseSensitive */ false, cwd);
            case 'Unix':
                return new mock_file_system_posix_1.MockFileSystemPosix(/* isCaseSensitive */ true, cwd);
            case 'Windows':
                return new mock_file_system_windows_1.MockFileSystemWindows(/* isCaseSensitive*/ false, cwd);
            case 'Native':
                return new mock_file_system_native_1.MockFileSystemNative(cwd);
            default:
                throw new Error('FileSystem not supported');
        }
    }
    function monkeyPatchTypeScript(os, fs) {
        ts.sys.directoryExists = function (path) {
            var absPath = fs.resolve(path);
            return fs.exists(absPath) && fs.stat(absPath).isDirectory();
        };
        ts.sys.fileExists = function (path) {
            var absPath = fs.resolve(path);
            return fs.exists(absPath) && fs.stat(absPath).isFile();
        };
        ts.sys.getCurrentDirectory = function () { return fs.pwd(); };
        ts.sys.getDirectories = getDirectories;
        ts.sys.readFile = fs.readFile.bind(fs);
        ts.sys.resolvePath = fs.resolve.bind(fs);
        ts.sys.writeFile = fs.writeFile.bind(fs);
        ts.sys.readDirectory = readDirectory;
        function getDirectories(path) {
            return fs.readdir(helpers_1.absoluteFrom(path)).filter(function (p) { return fs.stat(fs.resolve(path, p)).isDirectory(); });
        }
        function getFileSystemEntries(path) {
            var e_1, _a;
            var files = [];
            var directories = [];
            var absPath = fs.resolve(path);
            var entries = fs.readdir(absPath);
            try {
                for (var entries_1 = tslib_1.__values(entries), entries_1_1 = entries_1.next(); !entries_1_1.done; entries_1_1 = entries_1.next()) {
                    var entry = entries_1_1.value;
                    if (entry == '.' || entry === '..') {
                        continue;
                    }
                    var absPath_1 = fs.resolve(path, entry);
                    var stat = fs.stat(absPath_1);
                    if (stat.isDirectory()) {
                        directories.push(absPath_1);
                    }
                    else if (stat.isFile()) {
                        files.push(absPath_1);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (entries_1_1 && !entries_1_1.done && (_a = entries_1.return)) _a.call(entries_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return { files: files, directories: directories };
        }
        function realPath(path) {
            return fs.realpath(fs.resolve(path));
        }
        // Rather than completely re-implementing we are using the `ts.matchFiles` function,
        // which is internal to the `ts` namespace.
        var tsMatchFiles = ts.matchFiles;
        function readDirectory(path, extensions, excludes, includes, depth) {
            return tsMatchFiles(path, extensions, excludes, includes, fs.isCaseSensitive(), fs.pwd(), depth, getFileSystemEntries, realPath);
        }
    }
});
