// @angular/compiler-cli/src/ngtsc/incremental/src/strategy.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/incremental/src/strategy", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.PatchedProgramIncrementalBuildStrategy = exports.TrackedIncrementalBuildStrategy = exports.NoopIncrementalBuildStrategy = void 0;
    /**
     * A noop implementation of `IncrementalBuildStrategy` which neither returns nor tracks any
     * incremental data.
     */
    var NoopIncrementalBuildStrategy = /** @class */ (function () {
        function NoopIncrementalBuildStrategy() {
        }
        NoopIncrementalBuildStrategy.prototype.getIncrementalState = function () {
            return null;
        };
        NoopIncrementalBuildStrategy.prototype.setIncrementalState = function () { };
        NoopIncrementalBuildStrategy.prototype.toNextBuildStrategy = function () {
            return this;
        };
        return NoopIncrementalBuildStrategy;
    }());
    exports.NoopIncrementalBuildStrategy = NoopIncrementalBuildStrategy;
    /**
     * Tracks an `IncrementalDriver` within the strategy itself.
     */
    var TrackedIncrementalBuildStrategy = /** @class */ (function () {
        function TrackedIncrementalBuildStrategy() {
            this.state = null;
            this.isSet = false;
        }
        TrackedIncrementalBuildStrategy.prototype.getIncrementalState = function () {
            return this.state;
        };
        TrackedIncrementalBuildStrategy.prototype.setIncrementalState = function (state) {
            this.state = state;
            this.isSet = true;
        };
        TrackedIncrementalBuildStrategy.prototype.toNextBuildStrategy = function () {
            var strategy = new TrackedIncrementalBuildStrategy();
            // Only reuse state that was explicitly set via `setIncrementalState`.
            strategy.state = this.isSet ? this.state : null;
            return strategy;
        };
        return TrackedIncrementalBuildStrategy;
    }());
    exports.TrackedIncrementalBuildStrategy = TrackedIncrementalBuildStrategy;
    /**
     * Manages the `IncrementalDriver` associated with a `ts.Program` by monkey-patching it onto the
     * program under `SYM_INCREMENTAL_DRIVER`.
     */
    var PatchedProgramIncrementalBuildStrategy = /** @class */ (function () {
        function PatchedProgramIncrementalBuildStrategy() {
        }
        PatchedProgramIncrementalBuildStrategy.prototype.getIncrementalState = function (program) {
            var state = program[SYM_INCREMENTAL_STATE];
            if (state === undefined) {
                return null;
            }
            return state;
        };
        PatchedProgramIncrementalBuildStrategy.prototype.setIncrementalState = function (state, program) {
            program[SYM_INCREMENTAL_STATE] = state;
        };
        PatchedProgramIncrementalBuildStrategy.prototype.toNextBuildStrategy = function () {
            return this;
        };
        return PatchedProgramIncrementalBuildStrategy;
    }());
    exports.PatchedProgramIncrementalBuildStrategy = PatchedProgramIncrementalBuildStrategy;
    /**
     * Symbol under which the `IncrementalDriver` is stored on a `ts.Program`.
     *
     * The TS model of incremental compilation is based around reuse of a previous `ts.Program` in the
     * construction of a new one. The `NgCompiler` follows this abstraction - passing in a previous
     * `ts.Program` is sufficient to trigger incremental compilation. This previous `ts.Program` need
     * not be from an Angular compilation (that is, it need not have been created from `NgCompiler`).
     *
     * If it is, though, Angular can benefit from reusing previous analysis work. This reuse is managed
     * by the `IncrementalDriver`, which is inherited from the old program to the new program. To
     * support this behind the API of passing an old `ts.Program`, the `IncrementalDriver` is stored on
     * the `ts.Program` under this symbol.
     */
    var SYM_INCREMENTAL_STATE = Symbol('NgIncrementalState');
});
