// @ngrx/store/esm2022/src/meta-reducers/utils.mjs
export const RUNTIME_CHECK_URL = 'https://ngrx.io/guide/store/configuration/runtime-checks';
export function isUndefined(target) {
    return target === undefined;
}
export function isNull(target) {
    return target === null;
}
export function isArray(target) {
    return Array.isArray(target);
}
export function isString(target) {
    return typeof target === 'string';
}
export function isBoolean(target) {
    return typeof target === 'boolean';
}
export function isNumber(target) {
    return typeof target === 'number';
}
export function isObjectLike(target) {
    return typeof target === 'object' && target !== null;
}
export function isObject(target) {
    return isObjectLike(target) && !isArray(target);
}
export function isPlainObject(target) {
    if (!isObject(target)) {
        return false;
    }
    const targetPrototype = Object.getPrototypeOf(target);
    return targetPrototype === Object.prototype || targetPrototype === null;
}
export function isFunction(target) {
    return typeof target === 'function';
}
export function isComponent(target) {
    return isFunction(target) && target.hasOwnProperty('ɵcmp');
}
export function hasOwnProperty(target, propertyName) {
    return Object.prototype.hasOwnProperty.call(target, propertyName);
}
