// @angular/compiler-cli/linker/src/ast/typescript/typescript_ast_host.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/ast/typescript/typescript_ast_host", ["require", "exports", "tslib", "typescript", "@angular/compiler-cli/linker/src/fatal_linker_error", "@angular/compiler-cli/linker/src/ast/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TypeScriptAstHost = void 0;
    var tslib_1 = require("tslib");
    var ts = require("typescript");
    var fatal_linker_error_1 = require("@angular/compiler-cli/linker/src/fatal_linker_error");
    var utils_1 = require("@angular/compiler-cli/linker/src/ast/utils");
    /**
     * This implementation of `AstHost` is able to get information from TypeScript AST nodes.
     *
     * This host is not actually used at runtime in the current code.
     *
     * It is implemented here to ensure that the `AstHost` abstraction is not unfairly skewed towards
     * the Babel implementation. It could also provide a basis for a 3rd TypeScript compiler plugin to
     * do linking in the future.
     */
    var TypeScriptAstHost = /** @class */ (function () {
        function TypeScriptAstHost() {
            this.isStringLiteral = ts.isStringLiteral;
            this.isNumericLiteral = ts.isNumericLiteral;
            this.isArrayLiteral = ts.isArrayLiteralExpression;
            this.isObjectLiteral = ts.isObjectLiteralExpression;
            this.isCallExpression = ts.isCallExpression;
        }
        TypeScriptAstHost.prototype.getSymbolName = function (node) {
            if (ts.isIdentifier(node)) {
                return node.text;
            }
            else if (ts.isPropertyAccessExpression(node) && ts.isIdentifier(node.name)) {
                return node.name.text;
            }
            else {
                return null;
            }
        };
        TypeScriptAstHost.prototype.parseStringLiteral = function (str) {
            utils_1.assert(str, this.isStringLiteral, 'a string literal');
            return str.text;
        };
        TypeScriptAstHost.prototype.parseNumericLiteral = function (num) {
            utils_1.assert(num, this.isNumericLiteral, 'a numeric literal');
            return parseInt(num.text);
        };
        TypeScriptAstHost.prototype.isBooleanLiteral = function (node) {
            return isBooleanLiteral(node) || isMinifiedBooleanLiteral(node);
        };
        TypeScriptAstHost.prototype.parseBooleanLiteral = function (bool) {
            if (isBooleanLiteral(bool)) {
                return bool.kind === ts.SyntaxKind.TrueKeyword;
            }
            else if (isMinifiedBooleanLiteral(bool)) {
                return !(+bool.operand.text);
            }
            else {
                throw new fatal_linker_error_1.FatalLinkerError(bool, 'Unsupported syntax, expected a boolean literal.');
            }
        };
        TypeScriptAstHost.prototype.parseArrayLiteral = function (array) {
            utils_1.assert(array, this.isArrayLiteral, 'an array literal');
            return array.elements.map(function (element) {
                utils_1.assert(element, isNotEmptyElement, 'element in array not to be empty');
                utils_1.assert(element, isNotSpreadElement, 'element in array not to use spread syntax');
                return element;
            });
        };
        TypeScriptAstHost.prototype.parseObjectLiteral = function (obj) {
            var e_1, _a;
            utils_1.assert(obj, this.isObjectLiteral, 'an object literal');
            var result = new Map();
            try {
                for (var _b = tslib_1.__values(obj.properties), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var property = _c.value;
                    utils_1.assert(property, ts.isPropertyAssignment, 'a property assignment');
                    utils_1.assert(property.name, isPropertyName, 'a property name');
                    result.set(property.name.text, property.initializer);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return result;
        };
        TypeScriptAstHost.prototype.isFunctionExpression = function (node) {
            return ts.isFunctionExpression(node) || ts.isArrowFunction(node);
        };
        TypeScriptAstHost.prototype.parseReturnValue = function (fn) {
            utils_1.assert(fn, this.isFunctionExpression, 'a function');
            if (!ts.isBlock(fn.body)) {
                // it is a simple array function expression: `(...) => expr`
                return fn.body;
            }
            // it is a function (arrow or normal) with a body. E.g.:
            // * `(...) => { stmt; ... }`
            // * `function(...) { stmt; ... }`
            if (fn.body.statements.length !== 1) {
                throw new fatal_linker_error_1.FatalLinkerError(fn.body, 'Unsupported syntax, expected a function body with a single return statement.');
            }
            var stmt = fn.body.statements[0];
            utils_1.assert(stmt, ts.isReturnStatement, 'a function body with a single return statement');
            if (stmt.expression === undefined) {
                throw new fatal_linker_error_1.FatalLinkerError(stmt, 'Unsupported syntax, expected function to return a value.');
            }
            return stmt.expression;
        };
        TypeScriptAstHost.prototype.parseCallee = function (call) {
            utils_1.assert(call, ts.isCallExpression, 'a call expression');
            return call.expression;
        };
        TypeScriptAstHost.prototype.parseArguments = function (call) {
            utils_1.assert(call, ts.isCallExpression, 'a call expression');
            return call.arguments.map(function (arg) {
                utils_1.assert(arg, isNotSpreadElement, 'argument not to use spread syntax');
                return arg;
            });
        };
        TypeScriptAstHost.prototype.getRange = function (node) {
            var file = node.getSourceFile();
            if (file === undefined) {
                throw new fatal_linker_error_1.FatalLinkerError(node, 'Unable to read range for node - it is missing parent information.');
            }
            var startPos = node.getStart();
            var endPos = node.getEnd();
            var _a = ts.getLineAndCharacterOfPosition(file, startPos), startLine = _a.line, startCol = _a.character;
            return { startLine: startLine, startCol: startCol, startPos: startPos, endPos: endPos };
        };
        return TypeScriptAstHost;
    }());
    exports.TypeScriptAstHost = TypeScriptAstHost;
    /**
     * Return true if the expression does not represent an empty element in an array literal.
     * For example in `[,foo]` the first element is "empty".
     */
    function isNotEmptyElement(e) {
        return !ts.isOmittedExpression(e);
    }
    /**
     * Return true if the expression is not a spread element of an array literal.
     * For example in `[x, ...rest]` the `...rest` expression is a spread element.
     */
    function isNotSpreadElement(e) {
        return !ts.isSpreadElement(e);
    }
    /**
     * Return true if the expression can be considered a text based property name.
     */
    function isPropertyName(e) {
        return ts.isIdentifier(e) || ts.isStringLiteral(e) || ts.isNumericLiteral(e);
    }
    /**
     * Return true if the node is either `true` or `false` literals.
     */
    function isBooleanLiteral(node) {
        return node.kind === ts.SyntaxKind.TrueKeyword || node.kind === ts.SyntaxKind.FalseKeyword;
    }
    /**
     * Return true if the node is either `!0` or `!1`.
     */
    function isMinifiedBooleanLiteral(node) {
        return ts.isPrefixUnaryExpression(node) && node.operator === ts.SyntaxKind.ExclamationToken &&
            ts.isNumericLiteral(node.operand) && (node.operand.text === '0' || node.operand.text === '1');
    }
});
