// ngx-store/dist/src/index.js
import { NgModule } from '@angular/core';
import { CookiesStorageService, LocalStorageService, SessionStorageService } from './service';
import { SharedStorageService } from './service/shared-storage.service';
export { CookieStorage, LocalStorage, SessionStorage, SharedStorage } from './decorator/webstorage';
export { WebStorageService, CookiesStorageService, LocalStorageService, SessionStorageService, SharedStorageService } from './service';
export { WEBSTORAGE_CONFIG } from './config';
export { NgxStorageEvent } from './utility/storage/storage-event';
var WebStorageModule = /** @class */ (function () {
    function WebStorageModule() {
    }
    WebStorageModule.decorators = [
        { type: NgModule, args: [{
                    providers: [
                        LocalStorageService,
                        SessionStorageService,
                        CookiesStorageService,
                        SharedStorageService,
                    ]
                },] },
    ];
    /** @nocollapse */
    WebStorageModule.ctorParameters = function () { return []; };
    return WebStorageModule;
}());
export { WebStorageModule };
