// @angular/compiler-cli/linker/src/file_linker/emit_scopes/emit_scope.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/emit_scopes/emit_scope", ["require", "exports", "@angular/compiler", "@angular/compiler-cli/linker/src/linker_import_generator"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.EmitScope = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var compiler_1 = require("@angular/compiler");
    var linker_import_generator_1 = require("@angular/compiler-cli/linker/src/linker_import_generator");
    /**
     * This class represents (from the point of view of the `FileLinker`) the scope in which
     * statements and expressions related to a linked partial declaration will be emitted.
     *
     * It holds a copy of a `ConstantPool` that is used to capture any constant statements that need to
     * be emitted in this context.
     *
     * This implementation will emit the definition and the constant statements separately.
     */
    var EmitScope = /** @class */ (function () {
        function EmitScope(ngImport, translator) {
            this.ngImport = ngImport;
            this.translator = translator;
            this.constantPool = new compiler_1.ConstantPool();
        }
        /**
         * Translate the given Output AST definition expression into a generic `TExpression`.
         *
         * Use a `LinkerImportGenerator` to handle any imports in the definition.
         */
        EmitScope.prototype.translateDefinition = function (definition) {
            return this.translator.translateExpression(definition, new linker_import_generator_1.LinkerImportGenerator(this.ngImport));
        };
        /**
         * Return any constant statements that are shared between all uses of this `EmitScope`.
         */
        EmitScope.prototype.getConstantStatements = function () {
            var _this = this;
            var importGenerator = new linker_import_generator_1.LinkerImportGenerator(this.ngImport);
            return this.constantPool.statements.map(function (statement) { return _this.translator.translateStatement(statement, importGenerator); });
        };
        return EmitScope;
    }());
    exports.EmitScope = EmitScope;
});
