// @angular/compiler-cli/src/ngtsc/perf/src/recorder.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/// <reference types="node" />
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/perf/src/recorder", ["require", "exports", "@angular/compiler-cli/src/ngtsc/perf/src/api", "@angular/compiler-cli/src/ngtsc/perf/src/clock"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.DelegatingPerfRecorder = exports.ActivePerfRecorder = void 0;
    var api_1 = require("@angular/compiler-cli/src/ngtsc/perf/src/api");
    var clock_1 = require("@angular/compiler-cli/src/ngtsc/perf/src/clock");
    /**
     * A `PerfRecorder` that actively tracks performance statistics.
     */
    var ActivePerfRecorder = /** @class */ (function () {
        function ActivePerfRecorder(zeroTime) {
            this.zeroTime = zeroTime;
            this.currentPhase = api_1.PerfPhase.Unaccounted;
            this.currentPhaseEntered = this.zeroTime;
            this.counters = Array(api_1.PerfEvent.LAST).fill(0);
            this.phaseTime = Array(api_1.PerfPhase.LAST).fill(0);
            this.bytes = Array(api_1.PerfCheckpoint.LAST).fill(0);
            // Take an initial memory snapshot before any other compilation work begins.
            this.memory(api_1.PerfCheckpoint.Initial);
        }
        /**
         * Creates an `ActivePerfRecoder` with its zero point set to the current time.
         */
        ActivePerfRecorder.zeroedToNow = function () {
            return new ActivePerfRecorder(clock_1.mark());
        };
        ActivePerfRecorder.prototype.reset = function () {
            this.counters = Array(api_1.PerfEvent.LAST).fill(0);
            this.phaseTime = Array(api_1.PerfPhase.LAST).fill(0);
            this.bytes = Array(api_1.PerfCheckpoint.LAST).fill(0);
            this.zeroTime = clock_1.mark();
            this.currentPhase = api_1.PerfPhase.Unaccounted;
            this.currentPhaseEntered = this.zeroTime;
        };
        ActivePerfRecorder.prototype.memory = function (after) {
            this.bytes[after] = process.memoryUsage().heapUsed;
        };
        ActivePerfRecorder.prototype.phase = function (phase) {
            var previous = this.currentPhase;
            this.phaseTime[this.currentPhase] += clock_1.timeSinceInMicros(this.currentPhaseEntered);
            this.currentPhase = phase;
            this.currentPhaseEntered = clock_1.mark();
            return previous;
        };
        ActivePerfRecorder.prototype.inPhase = function (phase, fn) {
            var previousPhase = this.phase(phase);
            try {
                return fn();
            }
            finally {
                this.phase(previousPhase);
            }
        };
        ActivePerfRecorder.prototype.eventCount = function (counter, incrementBy) {
            if (incrementBy === void 0) { incrementBy = 1; }
            this.counters[counter] += incrementBy;
        };
        /**
         * Return the current performance metrics as a serializable object.
         */
        ActivePerfRecorder.prototype.finalize = function () {
            // Track the last segment of time spent in `this.currentPhase` in the time array.
            this.phase(api_1.PerfPhase.Unaccounted);
            var results = {
                events: {},
                phases: {},
                memory: {},
            };
            for (var i = 0; i < this.phaseTime.length; i++) {
                if (this.phaseTime[i] > 0) {
                    results.phases[api_1.PerfPhase[i]] = this.phaseTime[i];
                }
            }
            for (var i = 0; i < this.phaseTime.length; i++) {
                if (this.counters[i] > 0) {
                    results.events[api_1.PerfEvent[i]] = this.counters[i];
                }
            }
            for (var i = 0; i < this.bytes.length; i++) {
                if (this.bytes[i] > 0) {
                    results.memory[api_1.PerfCheckpoint[i]] = this.bytes[i];
                }
            }
            return results;
        };
        return ActivePerfRecorder;
    }());
    exports.ActivePerfRecorder = ActivePerfRecorder;
    /**
     * A `PerfRecorder` that delegates to a target `PerfRecorder` which can be updated later.
     *
     * `DelegatingPerfRecorder` is useful when a compiler class that needs a `PerfRecorder` can outlive
     * the current compilation. This is true for most compiler classes as resource-only changes reuse
     * the same `NgCompiler` for a new compilation.
     */
    var DelegatingPerfRecorder = /** @class */ (function () {
        function DelegatingPerfRecorder(target) {
            this.target = target;
        }
        DelegatingPerfRecorder.prototype.eventCount = function (counter, incrementBy) {
            this.target.eventCount(counter, incrementBy);
        };
        DelegatingPerfRecorder.prototype.phase = function (phase) {
            return this.target.phase(phase);
        };
        DelegatingPerfRecorder.prototype.inPhase = function (phase, fn) {
            // Note: this doesn't delegate to `this.target.inPhase` but instead is implemented manually here
            // to avoid adding an additional frame of noise to the stack when debugging.
            var previousPhase = this.target.phase(phase);
            try {
                return fn();
            }
            finally {
                this.target.phase(previousPhase);
            }
        };
        DelegatingPerfRecorder.prototype.memory = function (after) {
            this.target.memory(after);
        };
        DelegatingPerfRecorder.prototype.reset = function () {
            this.target.reset();
        };
        return DelegatingPerfRecorder;
    }());
    exports.DelegatingPerfRecorder = DelegatingPerfRecorder;
});
