// @angular/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory", ["require", "exports", "tslib", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.attachComments = exports.createTemplateTail = exports.createTemplateMiddle = exports.TypeScriptAstFactory = void 0;
    var tslib_1 = require("tslib");
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var ts = require("typescript");
    /**
     * Different optimizers use different annotations on a function or method call to indicate its pure
     * status.
     */
    var PureAnnotation;
    (function (PureAnnotation) {
        /**
         * Closure's annotation for purity is `@pureOrBreakMyCode`, but this needs to be in a semantic
         * (jsdoc) enabled comment. Thus, the actual comment text for Closure must include the `*` that
         * turns a `/*` comment into a `/**` comment, as well as surrounding whitespace.
         */
        PureAnnotation["CLOSURE"] = "* @pureOrBreakMyCode ";
        PureAnnotation["TERSER"] = "@__PURE__";
    })(PureAnnotation || (PureAnnotation = {}));
    var UNARY_OPERATORS = {
        '+': ts.SyntaxKind.PlusToken,
        '-': ts.SyntaxKind.MinusToken,
        '!': ts.SyntaxKind.ExclamationToken,
    };
    var BINARY_OPERATORS = {
        '&&': ts.SyntaxKind.AmpersandAmpersandToken,
        '>': ts.SyntaxKind.GreaterThanToken,
        '>=': ts.SyntaxKind.GreaterThanEqualsToken,
        '&': ts.SyntaxKind.AmpersandToken,
        '/': ts.SyntaxKind.SlashToken,
        '==': ts.SyntaxKind.EqualsEqualsToken,
        '===': ts.SyntaxKind.EqualsEqualsEqualsToken,
        '<': ts.SyntaxKind.LessThanToken,
        '<=': ts.SyntaxKind.LessThanEqualsToken,
        '-': ts.SyntaxKind.MinusToken,
        '%': ts.SyntaxKind.PercentToken,
        '*': ts.SyntaxKind.AsteriskToken,
        '!=': ts.SyntaxKind.ExclamationEqualsToken,
        '!==': ts.SyntaxKind.ExclamationEqualsEqualsToken,
        '||': ts.SyntaxKind.BarBarToken,
        '+': ts.SyntaxKind.PlusToken,
        '??': ts.SyntaxKind.QuestionQuestionToken,
    };
    var VAR_TYPES = {
        'const': ts.NodeFlags.Const,
        'let': ts.NodeFlags.Let,
        'var': ts.NodeFlags.None,
    };
    /**
     * A TypeScript flavoured implementation of the AstFactory.
     */
    var TypeScriptAstFactory = /** @class */ (function () {
        function TypeScriptAstFactory(annotateForClosureCompiler) {
            this.annotateForClosureCompiler = annotateForClosureCompiler;
            this.externalSourceFiles = new Map();
            this.attachComments = attachComments;
            this.createArrayLiteral = ts.createArrayLiteral;
            this.createConditional = ts.createConditional;
            this.createElementAccess = ts.createElementAccess;
            this.createExpressionStatement = ts.createExpressionStatement;
            this.createIdentifier = ts.createIdentifier;
            this.createParenthesizedExpression = ts.createParen;
            this.createPropertyAccess = ts.createPropertyAccess;
            this.createThrowStatement = ts.createThrow;
            this.createTypeOfExpression = ts.createTypeOf;
        }
        TypeScriptAstFactory.prototype.createAssignment = function (target, value) {
            return ts.createBinary(target, ts.SyntaxKind.EqualsToken, value);
        };
        TypeScriptAstFactory.prototype.createBinaryExpression = function (leftOperand, operator, rightOperand) {
            return ts.createBinary(leftOperand, BINARY_OPERATORS[operator], rightOperand);
        };
        TypeScriptAstFactory.prototype.createBlock = function (body) {
            return ts.createBlock(body);
        };
        TypeScriptAstFactory.prototype.createCallExpression = function (callee, args, pure) {
            var call = ts.createCall(callee, undefined, args);
            if (pure) {
                ts.addSyntheticLeadingComment(call, ts.SyntaxKind.MultiLineCommentTrivia, this.annotateForClosureCompiler ? PureAnnotation.CLOSURE : PureAnnotation.TERSER, 
                /* trailing newline */ false);
            }
            return call;
        };
        TypeScriptAstFactory.prototype.createFunctionDeclaration = function (functionName, parameters, body) {
            if (!ts.isBlock(body)) {
                throw new Error("Invalid syntax, expected a block, but got " + ts.SyntaxKind[body.kind] + ".");
            }
            return ts.createFunctionDeclaration(undefined, undefined, undefined, functionName, undefined, parameters.map(function (param) { return ts.createParameter(undefined, undefined, undefined, param); }), undefined, body);
        };
        TypeScriptAstFactory.prototype.createFunctionExpression = function (functionName, parameters, body) {
            if (!ts.isBlock(body)) {
                throw new Error("Invalid syntax, expected a block, but got " + ts.SyntaxKind[body.kind] + ".");
            }
            return ts.createFunctionExpression(undefined, undefined, functionName !== null && functionName !== void 0 ? functionName : undefined, undefined, parameters.map(function (param) { return ts.createParameter(undefined, undefined, undefined, param); }), undefined, body);
        };
        TypeScriptAstFactory.prototype.createIfStatement = function (condition, thenStatement, elseStatement) {
            return ts.createIf(condition, thenStatement, elseStatement !== null && elseStatement !== void 0 ? elseStatement : undefined);
        };
        TypeScriptAstFactory.prototype.createLiteral = function (value) {
            if (value === undefined) {
                return ts.createIdentifier('undefined');
            }
            else if (value === null) {
                return ts.createNull();
            }
            else {
                return ts.createLiteral(value);
            }
        };
        TypeScriptAstFactory.prototype.createNewExpression = function (expression, args) {
            return ts.createNew(expression, undefined, args);
        };
        TypeScriptAstFactory.prototype.createObjectLiteral = function (properties) {
            return ts.createObjectLiteral(properties.map(function (prop) { return ts.createPropertyAssignment(prop.quoted ? ts.createLiteral(prop.propertyName) :
                ts.createIdentifier(prop.propertyName), prop.value); }));
        };
        TypeScriptAstFactory.prototype.createReturnStatement = function (expression) {
            return ts.createReturn(expression !== null && expression !== void 0 ? expression : undefined);
        };
        TypeScriptAstFactory.prototype.createTaggedTemplate = function (tag, template) {
            var templateLiteral;
            var length = template.elements.length;
            var head = template.elements[0];
            if (length === 1) {
                templateLiteral = ts.createNoSubstitutionTemplateLiteral(head.cooked, head.raw);
            }
            else {
                var spans = [];
                // Create the middle parts
                for (var i = 1; i < length - 1; i++) {
                    var _a = template.elements[i], cooked = _a.cooked, raw = _a.raw, range = _a.range;
                    var middle = createTemplateMiddle(cooked, raw);
                    if (range !== null) {
                        this.setSourceMapRange(middle, range);
                    }
                    spans.push(ts.createTemplateSpan(template.expressions[i - 1], middle));
                }
                // Create the tail part
                var resolvedExpression = template.expressions[length - 2];
                var templatePart = template.elements[length - 1];
                var templateTail = createTemplateTail(templatePart.cooked, templatePart.raw);
                if (templatePart.range !== null) {
                    this.setSourceMapRange(templateTail, templatePart.range);
                }
                spans.push(ts.createTemplateSpan(resolvedExpression, templateTail));
                // Put it all together
                templateLiteral =
                    ts.createTemplateExpression(ts.createTemplateHead(head.cooked, head.raw), spans);
            }
            if (head.range !== null) {
                this.setSourceMapRange(templateLiteral, head.range);
            }
            return ts.createTaggedTemplate(tag, templateLiteral);
        };
        TypeScriptAstFactory.prototype.createUnaryExpression = function (operator, operand) {
            return ts.createPrefix(UNARY_OPERATORS[operator], operand);
        };
        TypeScriptAstFactory.prototype.createVariableDeclaration = function (variableName, initializer, type) {
            return ts.createVariableStatement(undefined, ts.createVariableDeclarationList([ts.createVariableDeclaration(variableName, undefined, initializer !== null && initializer !== void 0 ? initializer : undefined)], VAR_TYPES[type]));
        };
        TypeScriptAstFactory.prototype.setSourceMapRange = function (node, sourceMapRange) {
            if (sourceMapRange === null) {
                return node;
            }
            var url = sourceMapRange.url;
            if (!this.externalSourceFiles.has(url)) {
                this.externalSourceFiles.set(url, ts.createSourceMapSource(url, sourceMapRange.content, function (pos) { return pos; }));
            }
            var source = this.externalSourceFiles.get(url);
            ts.setSourceMapRange(node, { pos: sourceMapRange.start.offset, end: sourceMapRange.end.offset, source: source });
            return node;
        };
        return TypeScriptAstFactory;
    }());
    exports.TypeScriptAstFactory = TypeScriptAstFactory;
    // HACK: Use this in place of `ts.createTemplateMiddle()`.
    // Revert once https://github.com/microsoft/TypeScript/issues/35374 is fixed.
    function createTemplateMiddle(cooked, raw) {
        var node = ts.createTemplateHead(cooked, raw);
        node.kind = ts.SyntaxKind.TemplateMiddle;
        return node;
    }
    exports.createTemplateMiddle = createTemplateMiddle;
    // HACK: Use this in place of `ts.createTemplateTail()`.
    // Revert once https://github.com/microsoft/TypeScript/issues/35374 is fixed.
    function createTemplateTail(cooked, raw) {
        var node = ts.createTemplateHead(cooked, raw);
        node.kind = ts.SyntaxKind.TemplateTail;
        return node;
    }
    exports.createTemplateTail = createTemplateTail;
    /**
     * Attach the given `leadingComments` to the `statement` node.
     *
     * @param statement The statement that will have comments attached.
     * @param leadingComments The comments to attach to the statement.
     */
    function attachComments(statement, leadingComments) {
        var e_1, _a, e_2, _b;
        try {
            for (var leadingComments_1 = tslib_1.__values(leadingComments), leadingComments_1_1 = leadingComments_1.next(); !leadingComments_1_1.done; leadingComments_1_1 = leadingComments_1.next()) {
                var comment = leadingComments_1_1.value;
                var commentKind = comment.multiline ? ts.SyntaxKind.MultiLineCommentTrivia :
                    ts.SyntaxKind.SingleLineCommentTrivia;
                if (comment.multiline) {
                    ts.addSyntheticLeadingComment(statement, commentKind, comment.toString(), comment.trailingNewline);
                }
                else {
                    try {
                        for (var _c = (e_2 = void 0, tslib_1.__values(comment.toString().split('\n'))), _d = _c.next(); !_d.done; _d = _c.next()) {
                            var line = _d.value;
                            ts.addSyntheticLeadingComment(statement, commentKind, line, comment.trailingNewline);
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (leadingComments_1_1 && !leadingComments_1_1.done && (_a = leadingComments_1.return)) _a.call(leadingComments_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    exports.attachComments = attachComments;
});
