// @angular/compiler-cli/src/ngtsc/reflection/index.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/reflection", ["require", "exports", "tslib", "@angular/compiler-cli/src/ngtsc/reflection/src/host", "@angular/compiler-cli/src/ngtsc/reflection/src/type_to_value", "@angular/compiler-cli/src/ngtsc/reflection/src/typescript", "@angular/compiler-cli/src/ngtsc/reflection/src/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.isNamedVariableDeclaration = exports.isNamedFunctionDeclaration = exports.isNamedClassDeclaration = exports.reflectTypeEntityToDeclaration = exports.reflectObjectLiteral = exports.reflectNameOfDeclaration = exports.reflectIdentifierOfDeclaration = exports.filterToMembersWithDecorator = exports.TypeScriptReflectionHost = exports.typeNodeToValueExpr = void 0;
    var tslib_1 = require("tslib");
    tslib_1.__exportStar(require("@angular/compiler-cli/src/ngtsc/reflection/src/host"), exports);
    var type_to_value_1 = require("@angular/compiler-cli/src/ngtsc/reflection/src/type_to_value");
    Object.defineProperty(exports, "typeNodeToValueExpr", { enumerable: true, get: function () { return type_to_value_1.typeNodeToValueExpr; } });
    var typescript_1 = require("@angular/compiler-cli/src/ngtsc/reflection/src/typescript");
    Object.defineProperty(exports, "TypeScriptReflectionHost", { enumerable: true, get: function () { return typescript_1.TypeScriptReflectionHost; } });
    Object.defineProperty(exports, "filterToMembersWithDecorator", { enumerable: true, get: function () { return typescript_1.filterToMembersWithDecorator; } });
    Object.defineProperty(exports, "reflectIdentifierOfDeclaration", { enumerable: true, get: function () { return typescript_1.reflectIdentifierOfDeclaration; } });
    Object.defineProperty(exports, "reflectNameOfDeclaration", { enumerable: true, get: function () { return typescript_1.reflectNameOfDeclaration; } });
    Object.defineProperty(exports, "reflectObjectLiteral", { enumerable: true, get: function () { return typescript_1.reflectObjectLiteral; } });
    Object.defineProperty(exports, "reflectTypeEntityToDeclaration", { enumerable: true, get: function () { return typescript_1.reflectTypeEntityToDeclaration; } });
    var util_1 = require("@angular/compiler-cli/src/ngtsc/reflection/src/util");
    Object.defineProperty(exports, "isNamedClassDeclaration", { enumerable: true, get: function () { return util_1.isNamedClassDeclaration; } });
    Object.defineProperty(exports, "isNamedFunctionDeclaration", { enumerable: true, get: function () { return util_1.isNamedFunctionDeclaration; } });
    Object.defineProperty(exports, "isNamedVariableDeclaration", { enumerable: true, get: function () { return util_1.isNamedVariableDeclaration; } });
});
