// @angular/compiler-cli/linker/babel/src/ast/babel_ast_factory.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/babel/src/ast/babel_ast_factory", ["require", "exports", "@babel/types", "@angular/compiler-cli/linker"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.BabelAstFactory = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var t = require("@babel/types");
    var linker_1 = require("@angular/compiler-cli/linker");
    /**
     * A Babel flavored implementation of the AstFactory.
     */
    var BabelAstFactory = /** @class */ (function () {
        function BabelAstFactory(
        /** The absolute path to the source file being compiled. */
        sourceUrl) {
            this.sourceUrl = sourceUrl;
            this.createArrayLiteral = t.arrayExpression;
            this.createBlock = t.blockStatement;
            this.createConditional = t.conditionalExpression;
            this.createExpressionStatement = t.expressionStatement;
            this.createIdentifier = t.identifier;
            this.createIfStatement = t.ifStatement;
            this.createNewExpression = t.newExpression;
            this.createParenthesizedExpression = t.parenthesizedExpression;
            this.createReturnStatement = t.returnStatement;
            this.createThrowStatement = t.throwStatement;
            this.createUnaryExpression = t.unaryExpression;
        }
        BabelAstFactory.prototype.attachComments = function (statement, leadingComments) {
            // We must process the comments in reverse because `t.addComment()` will add new ones in front.
            for (var i = leadingComments.length - 1; i >= 0; i--) {
                var comment = leadingComments[i];
                t.addComment(statement, 'leading', comment.toString(), !comment.multiline);
            }
        };
        BabelAstFactory.prototype.createAssignment = function (target, value) {
            linker_1.assert(target, isLExpression, 'must be a left hand side expression');
            return t.assignmentExpression('=', target, value);
        };
        BabelAstFactory.prototype.createBinaryExpression = function (leftOperand, operator, rightOperand) {
            switch (operator) {
                case '&&':
                case '||':
                case '??':
                    return t.logicalExpression(operator, leftOperand, rightOperand);
                default:
                    return t.binaryExpression(operator, leftOperand, rightOperand);
            }
        };
        BabelAstFactory.prototype.createCallExpression = function (callee, args, pure) {
            var call = t.callExpression(callee, args);
            if (pure) {
                t.addComment(call, 'leading', ' @__PURE__ ', /* line */ false);
            }
            return call;
        };
        BabelAstFactory.prototype.createElementAccess = function (expression, element) {
            return t.memberExpression(expression, element, /* computed */ true);
        };
        BabelAstFactory.prototype.createFunctionDeclaration = function (functionName, parameters, body) {
            linker_1.assert(body, t.isBlockStatement, 'a block');
            return t.functionDeclaration(t.identifier(functionName), parameters.map(function (param) { return t.identifier(param); }), body);
        };
        BabelAstFactory.prototype.createFunctionExpression = function (functionName, parameters, body) {
            linker_1.assert(body, t.isBlockStatement, 'a block');
            var name = functionName !== null ? t.identifier(functionName) : null;
            return t.functionExpression(name, parameters.map(function (param) { return t.identifier(param); }), body);
        };
        BabelAstFactory.prototype.createLiteral = function (value) {
            if (typeof value === 'string') {
                return t.stringLiteral(value);
            }
            else if (typeof value === 'number') {
                return t.numericLiteral(value);
            }
            else if (typeof value === 'boolean') {
                return t.booleanLiteral(value);
            }
            else if (value === undefined) {
                return t.identifier('undefined');
            }
            else if (value === null) {
                return t.nullLiteral();
            }
            else {
                throw new Error("Invalid literal: " + value + " (" + typeof value + ")");
            }
        };
        BabelAstFactory.prototype.createObjectLiteral = function (properties) {
            return t.objectExpression(properties.map(function (prop) {
                var key = prop.quoted ? t.stringLiteral(prop.propertyName) : t.identifier(prop.propertyName);
                return t.objectProperty(key, prop.value);
            }));
        };
        BabelAstFactory.prototype.createPropertyAccess = function (expression, propertyName) {
            return t.memberExpression(expression, t.identifier(propertyName), /* computed */ false);
        };
        BabelAstFactory.prototype.createTaggedTemplate = function (tag, template) {
            var _this = this;
            var elements = template.elements.map(function (element, i) { return _this.setSourceMapRange(t.templateElement(element, i === template.elements.length - 1), element.range); });
            return t.taggedTemplateExpression(tag, t.templateLiteral(elements, template.expressions));
        };
        BabelAstFactory.prototype.createTypeOfExpression = function (expression) {
            return t.unaryExpression('typeof', expression);
        };
        BabelAstFactory.prototype.createVariableDeclaration = function (variableName, initializer, type) {
            return t.variableDeclaration(type, [t.variableDeclarator(t.identifier(variableName), initializer)]);
        };
        BabelAstFactory.prototype.setSourceMapRange = function (node, sourceMapRange) {
            if (sourceMapRange === null) {
                return node;
            }
            node.loc = {
                // Add in the filename so that we can map to external template files.
                // Note that Babel gets confused if you specify a filename when it is the original source
                // file. This happens when the template is inline, in which case just use `undefined`.
                filename: sourceMapRange.url !== this.sourceUrl ? sourceMapRange.url : undefined,
                start: {
                    line: sourceMapRange.start.line + 1,
                    column: sourceMapRange.start.column,
                },
                end: {
                    line: sourceMapRange.end.line + 1,
                    column: sourceMapRange.end.column,
                },
            }; // Needed because the Babel typings for `loc` don't include `filename`.
            node.start = sourceMapRange.start.offset;
            node.end = sourceMapRange.end.offset;
            return node;
        };
        return BabelAstFactory;
    }());
    exports.BabelAstFactory = BabelAstFactory;
    function isLExpression(expr) {
        // Some LVal types are not expressions, which prevents us from using `t.isLVal()`
        // directly with `assert()`.
        return t.isLVal(expr);
    }
});
