// @angular/compiler-cli/linker/babel/src/ast/babel_ast_host.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/babel/src/ast/babel_ast_host", ["require", "exports", "tslib", "@babel/types", "@angular/compiler-cli/linker"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.BabelAstHost = void 0;
    var tslib_1 = require("tslib");
    var t = require("@babel/types");
    var linker_1 = require("@angular/compiler-cli/linker");
    /**
     * This implementation of `AstHost` is able to get information from Babel AST nodes.
     */
    var BabelAstHost = /** @class */ (function () {
        function BabelAstHost() {
            this.isStringLiteral = t.isStringLiteral;
            this.isNumericLiteral = t.isNumericLiteral;
            this.isArrayLiteral = t.isArrayExpression;
            this.isObjectLiteral = t.isObjectExpression;
            this.isCallExpression = t.isCallExpression;
        }
        BabelAstHost.prototype.getSymbolName = function (node) {
            if (t.isIdentifier(node)) {
                return node.name;
            }
            else if (t.isMemberExpression(node) && t.isIdentifier(node.property)) {
                return node.property.name;
            }
            else {
                return null;
            }
        };
        BabelAstHost.prototype.parseStringLiteral = function (str) {
            linker_1.assert(str, t.isStringLiteral, 'a string literal');
            return str.value;
        };
        BabelAstHost.prototype.parseNumericLiteral = function (num) {
            linker_1.assert(num, t.isNumericLiteral, 'a numeric literal');
            return num.value;
        };
        BabelAstHost.prototype.isBooleanLiteral = function (bool) {
            return t.isBooleanLiteral(bool) || isMinifiedBooleanLiteral(bool);
        };
        BabelAstHost.prototype.parseBooleanLiteral = function (bool) {
            if (t.isBooleanLiteral(bool)) {
                return bool.value;
            }
            else if (isMinifiedBooleanLiteral(bool)) {
                return !bool.argument.value;
            }
            else {
                throw new linker_1.FatalLinkerError(bool, 'Unsupported syntax, expected a boolean literal.');
            }
        };
        BabelAstHost.prototype.parseArrayLiteral = function (array) {
            linker_1.assert(array, t.isArrayExpression, 'an array literal');
            return array.elements.map(function (element) {
                linker_1.assert(element, isNotEmptyElement, 'element in array not to be empty');
                linker_1.assert(element, isNotSpreadElement, 'element in array not to use spread syntax');
                return element;
            });
        };
        BabelAstHost.prototype.parseObjectLiteral = function (obj) {
            var e_1, _a;
            linker_1.assert(obj, t.isObjectExpression, 'an object literal');
            var result = new Map();
            try {
                for (var _b = tslib_1.__values(obj.properties), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var property = _c.value;
                    linker_1.assert(property, t.isObjectProperty, 'a property assignment');
                    linker_1.assert(property.value, t.isExpression, 'an expression');
                    linker_1.assert(property.key, isPropertyName, 'a property name');
                    var key = t.isIdentifier(property.key) ? property.key.name : property.key.value;
                    result.set(key, property.value);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return result;
        };
        BabelAstHost.prototype.isFunctionExpression = function (node) {
            return t.isFunction(node);
        };
        BabelAstHost.prototype.parseReturnValue = function (fn) {
            linker_1.assert(fn, this.isFunctionExpression, 'a function');
            if (!t.isBlockStatement(fn.body)) {
                // it is a simple array function expression: `(...) => expr`
                return fn.body;
            }
            // it is a function (arrow or normal) with a body. E.g.:
            // * `(...) => { stmt; ... }`
            // * `function(...) { stmt; ... }`
            if (fn.body.body.length !== 1) {
                throw new linker_1.FatalLinkerError(fn.body, 'Unsupported syntax, expected a function body with a single return statement.');
            }
            var stmt = fn.body.body[0];
            linker_1.assert(stmt, t.isReturnStatement, 'a function body with a single return statement');
            if (stmt.argument === null) {
                throw new linker_1.FatalLinkerError(stmt, 'Unsupported syntax, expected function to return a value.');
            }
            return stmt.argument;
        };
        BabelAstHost.prototype.parseCallee = function (call) {
            linker_1.assert(call, t.isCallExpression, 'a call expression');
            linker_1.assert(call.callee, t.isExpression, 'an expression');
            return call.callee;
        };
        BabelAstHost.prototype.parseArguments = function (call) {
            linker_1.assert(call, t.isCallExpression, 'a call expression');
            return call.arguments.map(function (arg) {
                linker_1.assert(arg, isNotSpreadArgument, 'argument not to use spread syntax');
                linker_1.assert(arg, t.isExpression, 'argument to be an expression');
                return arg;
            });
        };
        BabelAstHost.prototype.getRange = function (node) {
            if (node.loc == null || node.start === null || node.end === null) {
                throw new linker_1.FatalLinkerError(node, 'Unable to read range for node - it is missing location information.');
            }
            return {
                startLine: node.loc.start.line - 1,
                startCol: node.loc.start.column,
                startPos: node.start,
                endPos: node.end,
            };
        };
        return BabelAstHost;
    }());
    exports.BabelAstHost = BabelAstHost;
    /**
     * Return true if the expression does not represent an empty element in an array literal.
     * For example in `[,foo]` the first element is "empty".
     */
    function isNotEmptyElement(e) {
        return e !== null;
    }
    /**
     * Return true if the expression is not a spread element of an array literal.
     * For example in `[x, ...rest]` the `...rest` expression is a spread element.
     */
    function isNotSpreadElement(e) {
        return !t.isSpreadElement(e);
    }
    /**
     * Return true if the expression can be considered a text based property name.
     */
    function isPropertyName(e) {
        return t.isIdentifier(e) || t.isStringLiteral(e) || t.isNumericLiteral(e);
    }
    /**
     * Return true if the argument is not a spread element.
     */
    function isNotSpreadArgument(arg) {
        return !t.isSpreadElement(arg);
    }
    /**
     * Return true if the node is either `!0` or `!1`.
     */
    function isMinifiedBooleanLiteral(node) {
        return t.isUnaryExpression(node) && node.prefix && node.operator === '!' &&
            t.isNumericLiteral(node.argument) && (node.argument.value === 0 || node.argument.value === 1);
    }
});
