// @angular/compiler-cli/src/ngtsc/logging/testing/src/mock_logger.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/logging/testing/src/mock_logger", ["require", "exports", "@angular/compiler-cli/src/ngtsc/logging"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.MockLogger = void 0;
    var __1 = require("@angular/compiler-cli/src/ngtsc/logging");
    var MockLogger = /** @class */ (function () {
        function MockLogger(level) {
            if (level === void 0) { level = __1.LogLevel.info; }
            this.level = level;
            this.logs = {
                debug: [],
                info: [],
                warn: [],
                error: [],
            };
        }
        MockLogger.prototype.debug = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            this.logs.debug.push(args);
        };
        MockLogger.prototype.info = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            this.logs.info.push(args);
        };
        MockLogger.prototype.warn = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            this.logs.warn.push(args);
        };
        MockLogger.prototype.error = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            this.logs.error.push(args);
        };
        return MockLogger;
    }());
    exports.MockLogger = MockLogger;
});
