// apollo-angular/esm2022/testing/operation.mjs
import { Kind, OperationTypeNode } from 'graphql';
import { ApolloError } from '@apollo/client/core';
import { getMainDefinition } from '@apollo/client/utilities';
const isApolloError = (err) => err && err.hasOwnProperty('graphQLErrors');
export class TestOperation {
    operation;
    observer;
    constructor(operation, observer) {
        this.operation = operation;
        this.observer = observer;
    }
    flush(result) {
        if (isApolloError(result)) {
            this.observer.error(result);
        }
        else {
            const fetchResult = result ? { ...result } : result;
            this.observer.next(fetchResult);
            const definition = getMainDefinition(this.operation.query);
            if (definition.kind === Kind.OPERATION_DEFINITION &&
                definition.operation !== OperationTypeNode.SUBSCRIPTION) {
                this.complete();
            }
        }
    }
    complete() {
        this.observer.complete();
    }
    flushData(data) {
        this.flush({
            data,
        });
    }
    networkError(error) {
        const apolloError = new ApolloError({
            networkError: error,
        });
        this.flush(apolloError);
    }
    graphqlErrors(errors) {
        this.flush({
            errors,
        });
    }
}
