// @angular/core/esm2022/primitives/event-dispatch/src/bootstrap_global.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.dev/license
 */
import { addEvents, createEarlyJsactionData, getQueuedEventInfos, registerDispatcher, removeAllEventListeners, } from './earlyeventcontract';
/** Creates an `EarlyJsactionData`, adds events to it, and populates it on the window. */
export function bootstrapGlobalEarlyEventContract(bubbleEventTypes, captureEventTypes) {
    const earlyJsactionData = createEarlyJsactionData(window.document.documentElement);
    addEvents(earlyJsactionData, bubbleEventTypes);
    addEvents(earlyJsactionData, captureEventTypes, /* capture= */ true);
    window._ejsa = earlyJsactionData;
}
/** Get the queued `EventInfo` objects that were dispatched before a dispatcher was registered. */
export function getGlobalQueuedEventInfos() {
    return getQueuedEventInfos(window._ejsa);
}
/** Registers a dispatcher function on the `EarlyJsactionData` present on the window. */
export function registerGlobalDispatcher(restriction, dispatcher) {
    registerDispatcher(window._ejsa, dispatcher);
}
/** Removes all event listener handlers. */
export function removeAllGlobalEventListeners() {
    removeAllEventListeners(window._ejsa);
}
/** Removes the global early event contract. */
export function clearGlobalEarlyEventContract() {
    window._ejsa = undefined;
}
