// @angular/compiler-cli/src/ngtsc/shims/index.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/shims", ["require", "exports", "@angular/compiler-cli/src/ngtsc/shims/src/adapter", "@angular/compiler-cli/src/ngtsc/shims/src/expando", "@angular/compiler-cli/src/ngtsc/shims/src/factory_generator", "@angular/compiler-cli/src/ngtsc/shims/src/reference_tagger", "@angular/compiler-cli/src/ngtsc/shims/src/summary_generator"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.SummaryGenerator = exports.ShimReferenceTagger = exports.generatedFactoryTransform = exports.FactoryGenerator = exports.untagTsFile = exports.untagAllTsFiles = exports.sfExtensionData = exports.retagTsFile = exports.retagAllTsFiles = exports.isShim = exports.copyFileShimData = exports.ShimAdapter = void 0;
    /// <reference types="node" />
    var adapter_1 = require("@angular/compiler-cli/src/ngtsc/shims/src/adapter");
    Object.defineProperty(exports, "ShimAdapter", { enumerable: true, get: function () { return adapter_1.ShimAdapter; } });
    var expando_1 = require("@angular/compiler-cli/src/ngtsc/shims/src/expando");
    Object.defineProperty(exports, "copyFileShimData", { enumerable: true, get: function () { return expando_1.copyFileShimData; } });
    Object.defineProperty(exports, "isShim", { enumerable: true, get: function () { return expando_1.isShim; } });
    Object.defineProperty(exports, "retagAllTsFiles", { enumerable: true, get: function () { return expando_1.retagAllTsFiles; } });
    Object.defineProperty(exports, "retagTsFile", { enumerable: true, get: function () { return expando_1.retagTsFile; } });
    Object.defineProperty(exports, "sfExtensionData", { enumerable: true, get: function () { return expando_1.sfExtensionData; } });
    Object.defineProperty(exports, "untagAllTsFiles", { enumerable: true, get: function () { return expando_1.untagAllTsFiles; } });
    Object.defineProperty(exports, "untagTsFile", { enumerable: true, get: function () { return expando_1.untagTsFile; } });
    var factory_generator_1 = require("@angular/compiler-cli/src/ngtsc/shims/src/factory_generator");
    Object.defineProperty(exports, "FactoryGenerator", { enumerable: true, get: function () { return factory_generator_1.FactoryGenerator; } });
    Object.defineProperty(exports, "generatedFactoryTransform", { enumerable: true, get: function () { return factory_generator_1.generatedFactoryTransform; } });
    var reference_tagger_1 = require("@angular/compiler-cli/src/ngtsc/shims/src/reference_tagger");
    Object.defineProperty(exports, "ShimReferenceTagger", { enumerable: true, get: function () { return reference_tagger_1.ShimReferenceTagger; } });
    var summary_generator_1 = require("@angular/compiler-cli/src/ngtsc/shims/src/summary_generator");
    Object.defineProperty(exports, "SummaryGenerator", { enumerable: true, get: function () { return summary_generator_1.SummaryGenerator; } });
});
