// @angular/compiler-cli/linker/src/file_linker/file_linker.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/file_linker", ["require", "exports", "tslib", "@angular/compiler-cli/linker/src/ast/ast_value", "@angular/compiler-cli/linker/src/file_linker/emit_scopes/emit_scope", "@angular/compiler-cli/linker/src/file_linker/emit_scopes/iife_emit_scope", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_linker_selector"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.FileLinker = exports.NO_STATEMENTS = void 0;
    var tslib_1 = require("tslib");
    var ast_value_1 = require("@angular/compiler-cli/linker/src/ast/ast_value");
    var emit_scope_1 = require("@angular/compiler-cli/linker/src/file_linker/emit_scopes/emit_scope");
    var iife_emit_scope_1 = require("@angular/compiler-cli/linker/src/file_linker/emit_scopes/iife_emit_scope");
    var partial_linker_selector_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_linker_selector");
    exports.NO_STATEMENTS = [];
    /**
     * This class is responsible for linking all the partial declarations found in a single file.
     */
    var FileLinker = /** @class */ (function () {
        function FileLinker(linkerEnvironment, sourceUrl, code) {
            this.linkerEnvironment = linkerEnvironment;
            this.emitScopes = new Map();
            this.linkerSelector = new partial_linker_selector_1.PartialLinkerSelector(partial_linker_selector_1.createLinkerMap(this.linkerEnvironment, sourceUrl, code), this.linkerEnvironment.logger, this.linkerEnvironment.options.unknownDeclarationVersionHandling);
        }
        /**
         * Return true if the given callee name matches a partial declaration that can be linked.
         */
        FileLinker.prototype.isPartialDeclaration = function (calleeName) {
            return this.linkerSelector.supportsDeclaration(calleeName);
        };
        /**
         * Link the metadata extracted from the args of a call to a partial declaration function.
         *
         * The `declarationScope` is used to determine the scope and strategy of emission of the linked
         * definition and any shared constant statements.
         *
         * @param declarationFn the name of the function used to declare the partial declaration - e.g.
         *     `ɵɵngDeclareDirective`.
         * @param args the arguments passed to the declaration function, should be a single object that
         *     corresponds to the `R3DeclareDirectiveMetadata` or `R3DeclareComponentMetadata` interfaces.
         * @param declarationScope the scope that contains this call to the declaration function.
         */
        FileLinker.prototype.linkPartialDeclaration = function (declarationFn, args, declarationScope) {
            if (args.length !== 1) {
                throw new Error("Invalid function call: It should have only a single object literal argument, but contained " + args.length + ".");
            }
            var metaObj = ast_value_1.AstObject.parse(args[0], this.linkerEnvironment.host);
            var ngImport = metaObj.getNode('ngImport');
            var emitScope = this.getEmitScope(ngImport, declarationScope);
            var minVersion = metaObj.getString('minVersion');
            var version = metaObj.getString('version');
            var linker = this.linkerSelector.getLinker(declarationFn, minVersion, version);
            var definition = linker.linkPartialDeclaration(emitScope.constantPool, metaObj);
            return emitScope.translateDefinition(definition);
        };
        /**
         * Return all the shared constant statements and their associated constant scope references, so
         * that they can be inserted into the source code.
         */
        FileLinker.prototype.getConstantStatements = function () {
            var e_1, _a;
            var results = [];
            try {
                for (var _b = tslib_1.__values(this.emitScopes.entries()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var _d = tslib_1.__read(_c.value, 2), constantScope = _d[0], emitScope = _d[1];
                    var statements = emitScope.getConstantStatements();
                    results.push({ constantScope: constantScope, statements: statements });
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return results;
        };
        FileLinker.prototype.getEmitScope = function (ngImport, declarationScope) {
            var constantScope = declarationScope.getConstantScopeRef(ngImport);
            if (constantScope === null) {
                // There is no constant scope so we will emit extra statements into the definition IIFE.
                return new iife_emit_scope_1.IifeEmitScope(ngImport, this.linkerEnvironment.translator, this.linkerEnvironment.factory);
            }
            if (!this.emitScopes.has(constantScope)) {
                this.emitScopes.set(constantScope, new emit_scope_1.EmitScope(ngImport, this.linkerEnvironment.translator));
            }
            return this.emitScopes.get(constantScope);
        };
        return FileLinker;
    }());
    exports.FileLinker = FileLinker;
});
