// @angular/compiler-cli/src/ngtsc/cycles/src/imports.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/cycles/src/imports", ["require", "exports", "tslib", "typescript", "@angular/compiler-cli/src/ngtsc/perf"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ImportGraph = void 0;
    var tslib_1 = require("tslib");
    var ts = require("typescript");
    var perf_1 = require("@angular/compiler-cli/src/ngtsc/perf");
    /**
     * A cached graph of imports in the `ts.Program`.
     *
     * The `ImportGraph` keeps track of dependencies (imports) of individual `ts.SourceFile`s. Only
     * dependencies within the same program are tracked; imports into packages on NPM are not.
     */
    var ImportGraph = /** @class */ (function () {
        function ImportGraph(checker, perf) {
            this.checker = checker;
            this.perf = perf;
            this.map = new Map();
        }
        /**
         * List the direct (not transitive) imports of a given `ts.SourceFile`.
         *
         * This operation is cached.
         */
        ImportGraph.prototype.importsOf = function (sf) {
            if (!this.map.has(sf)) {
                this.map.set(sf, this.scanImports(sf));
            }
            return this.map.get(sf);
        };
        /**
         * Lists the transitive imports of a given `ts.SourceFile`.
         */
        ImportGraph.prototype.transitiveImportsOf = function (sf) {
            var imports = new Set();
            this.transitiveImportsOfHelper(sf, imports);
            return imports;
        };
        ImportGraph.prototype.transitiveImportsOfHelper = function (sf, results) {
            var _this = this;
            if (results.has(sf)) {
                return;
            }
            results.add(sf);
            this.importsOf(sf).forEach(function (imported) {
                _this.transitiveImportsOfHelper(imported, results);
            });
        };
        /**
         * Find an import path from the `start` SourceFile to the `end` SourceFile.
         *
         * This function implements a breadth first search that results in finding the
         * shortest path between the `start` and `end` points.
         *
         * @param start the starting point of the path.
         * @param end the ending point of the path.
         * @returns an array of source files that connect the `start` and `end` source files, or `null` if
         *     no path could be found.
         */
        ImportGraph.prototype.findPath = function (start, end) {
            var e_1, _a;
            if (start === end) {
                // Escape early for the case where `start` and `end` are the same.
                return [start];
            }
            var found = new Set([start]);
            var queue = [new Found(start, null)];
            while (queue.length > 0) {
                var current = queue.shift();
                var imports = this.importsOf(current.sourceFile);
                try {
                    for (var imports_1 = (e_1 = void 0, tslib_1.__values(imports)), imports_1_1 = imports_1.next(); !imports_1_1.done; imports_1_1 = imports_1.next()) {
                        var importedFile = imports_1_1.value;
                        if (!found.has(importedFile)) {
                            var next = new Found(importedFile, current);
                            if (next.sourceFile === end) {
                                // We have hit the target `end` path so we can stop here.
                                return next.toPath();
                            }
                            found.add(importedFile);
                            queue.push(next);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (imports_1_1 && !imports_1_1.done && (_a = imports_1.return)) _a.call(imports_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            return null;
        };
        /**
         * Add a record of an import from `sf` to `imported`, that's not present in the original
         * `ts.Program` but will be remembered by the `ImportGraph`.
         */
        ImportGraph.prototype.addSyntheticImport = function (sf, imported) {
            if (isLocalFile(imported)) {
                this.importsOf(sf).add(imported);
            }
        };
        ImportGraph.prototype.scanImports = function (sf) {
            var _this = this;
            return this.perf.inPhase(perf_1.PerfPhase.CycleDetection, function () {
                var e_2, _a;
                var imports = new Set();
                try {
                    // Look through the source file for import and export statements.
                    for (var _b = tslib_1.__values(sf.statements), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var stmt = _c.value;
                        if ((!ts.isImportDeclaration(stmt) && !ts.isExportDeclaration(stmt)) ||
                            stmt.moduleSpecifier === undefined) {
                            continue;
                        }
                        var symbol = _this.checker.getSymbolAtLocation(stmt.moduleSpecifier);
                        if (symbol === undefined || symbol.valueDeclaration === undefined) {
                            // No symbol could be found to skip over this import/export.
                            continue;
                        }
                        var moduleFile = symbol.valueDeclaration;
                        if (ts.isSourceFile(moduleFile) && isLocalFile(moduleFile)) {
                            // Record this local import.
                            imports.add(moduleFile);
                        }
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
                return imports;
            });
        };
        return ImportGraph;
    }());
    exports.ImportGraph = ImportGraph;
    function isLocalFile(sf) {
        return !sf.isDeclarationFile;
    }
    /**
     * A helper class to track which SourceFiles are being processed when searching for a path in
     * `getPath()` above.
     */
    var Found = /** @class */ (function () {
        function Found(sourceFile, parent) {
            this.sourceFile = sourceFile;
            this.parent = parent;
        }
        /**
         * Back track through this found SourceFile and its ancestors to generate an array of
         * SourceFiles that form am import path between two SourceFiles.
         */
        Found.prototype.toPath = function () {
            var array = [];
            var current = this;
            while (current !== null) {
                array.push(current.sourceFile);
                current = current.parent;
            }
            // Pushing and then reversing, O(n), rather than unshifting repeatedly, O(n^2), avoids
            // manipulating the array on every iteration: https://stackoverflow.com/a/26370620
            return array.reverse();
        };
        return Found;
    }());
});
