// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_linker_selector.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_linker_selector", ["require", "exports", "tslib", "semver", "@angular/compiler-cli/linker/src/file_linker/get_source_file", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_class_metadata_linker_1", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_component_linker_1", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_factory_linker_1", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injectable_linker_1", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injector_linker_1", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_ng_module_linker_1", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_pipe_linker_1"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.PartialLinkerSelector = exports.createLinkerMap = exports.declarationFunctions = exports.ɵɵngDeclarePipe = exports.ɵɵngDeclareNgModule = exports.ɵɵngDeclareInjector = exports.ɵɵngDeclareInjectable = exports.ɵɵngDeclareFactory = exports.ɵɵngDeclareComponent = exports.ɵɵngDeclareClassMetadata = exports.ɵɵngDeclareDirective = void 0;
    var tslib_1 = require("tslib");
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var semver_1 = require("semver");
    var get_source_file_1 = require("@angular/compiler-cli/linker/src/file_linker/get_source_file");
    var partial_class_metadata_linker_1_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_class_metadata_linker_1");
    var partial_component_linker_1_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_component_linker_1");
    var partial_directive_linker_1_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1");
    var partial_factory_linker_1_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_factory_linker_1");
    var partial_injectable_linker_1_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injectable_linker_1");
    var partial_injector_linker_1_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injector_linker_1");
    var partial_ng_module_linker_1_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_ng_module_linker_1");
    var partial_pipe_linker_1_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_pipe_linker_1");
    exports.ɵɵngDeclareDirective = 'ɵɵngDeclareDirective';
    exports.ɵɵngDeclareClassMetadata = 'ɵɵngDeclareClassMetadata';
    exports.ɵɵngDeclareComponent = 'ɵɵngDeclareComponent';
    exports.ɵɵngDeclareFactory = 'ɵɵngDeclareFactory';
    exports.ɵɵngDeclareInjectable = 'ɵɵngDeclareInjectable';
    exports.ɵɵngDeclareInjector = 'ɵɵngDeclareInjector';
    exports.ɵɵngDeclareNgModule = 'ɵɵngDeclareNgModule';
    exports.ɵɵngDeclarePipe = 'ɵɵngDeclarePipe';
    exports.declarationFunctions = [
        exports.ɵɵngDeclareDirective, exports.ɵɵngDeclareClassMetadata, exports.ɵɵngDeclareComponent, exports.ɵɵngDeclareFactory,
        exports.ɵɵngDeclareInjectable, exports.ɵɵngDeclareInjector, exports.ɵɵngDeclareNgModule, exports.ɵɵngDeclarePipe
    ];
    /**
     * Create a mapping between partial-declaration call name and collections of partial-linkers.
     *
     * Each collection of partial-linkers will contain a version range that will be matched against the
     * `minVersion` of the partial-declaration. (Additionally, a partial-linker may modify its behaviour
     * internally based on the `version` property of the declaration.)
     *
     * Versions should be sorted in ascending order. The most recent partial-linker will be used as the
     * fallback linker if none of the other version ranges match. For example:
     *
     * ```
     * {range: getRange('<=', '13.0.0'), linker PartialDirectiveLinkerVersion2(...) },
     * {range: getRange('<=', '13.1.0'), linker PartialDirectiveLinkerVersion3(...) },
     * {range: getRange('<=', '14.0.0'), linker PartialDirectiveLinkerVersion4(...) },
     * {range: LATEST_VERSION_RANGE, linker: new PartialDirectiveLinkerVersion1(...)},
     * ```
     *
     * If the `LATEST_VERSION_RANGE` is `<=15.0.0` then the fallback linker would be
     * `PartialDirectiveLinkerVersion1` for any version greater than `15.0.0`.
     *
     * When there is a change to a declaration interface that requires a new partial-linker, the
     * `minVersion` of the partial-declaration should be updated, the new linker implementation should
     * be added to the end of the collection, and the version of the previous linker should be updated.
     */
    function createLinkerMap(environment, sourceUrl, code) {
        var linkers = new Map();
        var LATEST_VERSION_RANGE = getRange('<=', '12.0.2');
        linkers.set(exports.ɵɵngDeclareDirective, [
            { range: LATEST_VERSION_RANGE, linker: new partial_directive_linker_1_1.PartialDirectiveLinkerVersion1(sourceUrl, code) },
        ]);
        linkers.set(exports.ɵɵngDeclareClassMetadata, [
            { range: LATEST_VERSION_RANGE, linker: new partial_class_metadata_linker_1_1.PartialClassMetadataLinkerVersion1() },
        ]);
        linkers.set(exports.ɵɵngDeclareComponent, [
            {
                range: LATEST_VERSION_RANGE,
                linker: new partial_component_linker_1_1.PartialComponentLinkerVersion1(get_source_file_1.createGetSourceFile(sourceUrl, code, environment.sourceFileLoader), sourceUrl, code)
            },
        ]);
        linkers.set(exports.ɵɵngDeclareFactory, [
            { range: LATEST_VERSION_RANGE, linker: new partial_factory_linker_1_1.PartialFactoryLinkerVersion1() },
        ]);
        linkers.set(exports.ɵɵngDeclareInjectable, [
            { range: LATEST_VERSION_RANGE, linker: new partial_injectable_linker_1_1.PartialInjectableLinkerVersion1() },
        ]);
        linkers.set(exports.ɵɵngDeclareInjector, [
            { range: LATEST_VERSION_RANGE, linker: new partial_injector_linker_1_1.PartialInjectorLinkerVersion1() },
        ]);
        linkers.set(exports.ɵɵngDeclareNgModule, [
            {
                range: LATEST_VERSION_RANGE,
                linker: new partial_ng_module_linker_1_1.PartialNgModuleLinkerVersion1(environment.options.linkerJitMode)
            },
        ]);
        linkers.set(exports.ɵɵngDeclarePipe, [
            { range: LATEST_VERSION_RANGE, linker: new partial_pipe_linker_1_1.PartialPipeLinkerVersion1() },
        ]);
        return linkers;
    }
    exports.createLinkerMap = createLinkerMap;
    /**
     * A helper that selects the appropriate `PartialLinker` for a given declaration.
     *
     * The selection is made from a database of linker instances, chosen if their given semver range
     * satisfies the `minVersion` of the partial declaration to be linked.
     *
     * Note that the ranges are checked in order, and the first matching range will be selected. So
     * ranges should be most restrictive first. In practice, since ranges are always `<=X.Y.Z` this
     * means that ranges should be in ascending order.
     *
     * Note that any "pre-release" versions are stripped from ranges. Therefore if a `minVersion` is
     * `11.1.0-next.1` then this would match `11.1.0-next.2` and also `12.0.0-next.1`. (This is
     * different to standard semver range checking, where pre-release versions do not cross full version
     * boundaries.)
     */
    var PartialLinkerSelector = /** @class */ (function () {
        function PartialLinkerSelector(linkers, logger, unknownDeclarationVersionHandling) {
            this.linkers = linkers;
            this.logger = logger;
            this.unknownDeclarationVersionHandling = unknownDeclarationVersionHandling;
        }
        /**
         * Returns true if there are `PartialLinker` classes that can handle functions with this name.
         */
        PartialLinkerSelector.prototype.supportsDeclaration = function (functionName) {
            return this.linkers.has(functionName);
        };
        /**
         * Returns the `PartialLinker` that can handle functions with the given name and version.
         * Throws an error if there is none.
         */
        PartialLinkerSelector.prototype.getLinker = function (functionName, minVersion, version) {
            var e_1, _a;
            if (!this.linkers.has(functionName)) {
                throw new Error("Unknown partial declaration function " + functionName + ".");
            }
            var linkerRanges = this.linkers.get(functionName);
            if (version === '12.0.2') {
                // Special case if the `version` is the same as the current compiler version.
                // This helps with compliance tests where the version placeholders have not been replaced.
                return linkerRanges[linkerRanges.length - 1].linker;
            }
            var declarationRange = getRange('>=', minVersion);
            try {
                for (var linkerRanges_1 = tslib_1.__values(linkerRanges), linkerRanges_1_1 = linkerRanges_1.next(); !linkerRanges_1_1.done; linkerRanges_1_1 = linkerRanges_1.next()) {
                    var _b = linkerRanges_1_1.value, linkerRange = _b.range, linker = _b.linker;
                    if (semver_1.intersects(declarationRange, linkerRange)) {
                        return linker;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (linkerRanges_1_1 && !linkerRanges_1_1.done && (_a = linkerRanges_1.return)) _a.call(linkerRanges_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            var message = "This application depends upon a library published using Angular version " + version + ", " +
                ("which requires Angular version " + minVersion + " or newer to work correctly.\n") +
                "Consider upgrading your application to use a more recent version of Angular.";
            if (this.unknownDeclarationVersionHandling === 'error') {
                throw new Error(message);
            }
            else if (this.unknownDeclarationVersionHandling === 'warn') {
                this.logger.warn(message + "\nAttempting to continue using this version of Angular.");
            }
            // No linker was matched for this declaration, so just use the most recent one.
            return linkerRanges[linkerRanges.length - 1].linker;
        };
        return PartialLinkerSelector;
    }());
    exports.PartialLinkerSelector = PartialLinkerSelector;
    /**
     * Compute a semver Range from the `version` and comparator.
     *
     * The range is computed as any version greater/less than or equal to the given `versionStr`
     * depending upon the `comparator` (ignoring any prerelease versions).
     *
     * @param comparator a string that determines whether the version specifies a minimum or a maximum
     *     range.
     * @param versionStr the version given in the partial declaration
     * @returns A semver range for the provided `version` and comparator.
     */
    function getRange(comparator, versionStr) {
        var version = new semver_1.SemVer(versionStr);
        // Wipe out any prerelease versions
        version.prerelease = [];
        return new semver_1.Range("" + comparator + version.format());
    }
});
