// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_ng_module_linker_1.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_ng_module_linker_1", ["require", "exports", "@angular/compiler", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.toR3NgModuleMeta = exports.PartialNgModuleLinkerVersion1 = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var compiler_1 = require("@angular/compiler");
    var util_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/util");
    /**
     * A `PartialLinker` that is designed to process `ɵɵngDeclareNgModule()` call expressions.
     */
    var PartialNgModuleLinkerVersion1 = /** @class */ (function () {
        function PartialNgModuleLinkerVersion1(
        /**
         * If true then emit the additional declarations, imports, exports, etc in the NgModule
         * definition. These are only used by JIT compilation.
         */
        emitInline) {
            this.emitInline = emitInline;
        }
        PartialNgModuleLinkerVersion1.prototype.linkPartialDeclaration = function (constantPool, metaObj) {
            var meta = toR3NgModuleMeta(metaObj, this.emitInline);
            var def = compiler_1.compileNgModule(meta);
            return def.expression;
        };
        return PartialNgModuleLinkerVersion1;
    }());
    exports.PartialNgModuleLinkerVersion1 = PartialNgModuleLinkerVersion1;
    /**
     * Derives the `R3NgModuleMetadata` structure from the AST object.
     */
    function toR3NgModuleMeta(metaObj, emitInline) {
        var wrappedType = metaObj.getOpaque('type');
        var meta = {
            type: util_1.wrapReference(wrappedType),
            internalType: wrappedType,
            adjacentType: wrappedType,
            bootstrap: [],
            declarations: [],
            imports: [],
            exports: [],
            emitInline: emitInline,
            containsForwardDecls: false,
            schemas: [],
            id: metaObj.has('id') ? metaObj.getOpaque('id') : null,
        };
        // Each of `bootstrap`, `declarations`, `imports` and `exports` are normally an array. But if any
        // of the references are not yet declared, then the arrays must be wrapped in a function to
        // prevent errors at runtime when accessing the values.
        // The following blocks of code will unwrap the arrays from such functions, because
        // `R3NgModuleMetadata` expects arrays of `R3Reference` objects.
        // Further, since the `ɵɵdefineNgModule()` will also suffer from the forward declaration problem,
        // we must update the `containsForwardDecls` property if a function wrapper was found.
        if (metaObj.has('bootstrap')) {
            var bootstrap = metaObj.getValue('bootstrap');
            if (bootstrap.isFunction()) {
                meta.containsForwardDecls = true;
                meta.bootstrap = wrapReferences(unwrapForwardRefs(bootstrap));
            }
            else
                meta.bootstrap = wrapReferences(bootstrap);
        }
        if (metaObj.has('declarations')) {
            var declarations = metaObj.getValue('declarations');
            if (declarations.isFunction()) {
                meta.containsForwardDecls = true;
                meta.declarations = wrapReferences(unwrapForwardRefs(declarations));
            }
            else
                meta.declarations = wrapReferences(declarations);
        }
        if (metaObj.has('imports')) {
            var imports = metaObj.getValue('imports');
            if (imports.isFunction()) {
                meta.containsForwardDecls = true;
                meta.imports = wrapReferences(unwrapForwardRefs(imports));
            }
            else
                meta.imports = wrapReferences(imports);
        }
        if (metaObj.has('exports')) {
            var exports_1 = metaObj.getValue('exports');
            if (exports_1.isFunction()) {
                meta.containsForwardDecls = true;
                meta.exports = wrapReferences(unwrapForwardRefs(exports_1));
            }
            else
                meta.exports = wrapReferences(exports_1);
        }
        if (metaObj.has('schemas')) {
            var schemas = metaObj.getValue('schemas');
            meta.schemas = wrapReferences(schemas);
        }
        return meta;
    }
    exports.toR3NgModuleMeta = toR3NgModuleMeta;
    /**
     * Extract an array from the body of the function.
     *
     * If `field` is `function() { return [exp1, exp2, exp3]; }` then we return `[exp1, exp2, exp3]`.
     *
     */
    function unwrapForwardRefs(field) {
        return field.getFunctionReturnValue();
    }
    /**
     * Wrap the array of expressions into an array of R3 references.
     */
    function wrapReferences(values) {
        return values.getArray().map(function (i) { return util_1.wrapReference(i.getOpaque()); });
    }
});
