// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1", ["require", "exports", "@angular/compiler", "@angular/compiler-cli/linker/src/fatal_linker_error", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.createSourceSpan = exports.toR3DirectiveMeta = exports.PartialDirectiveLinkerVersion1 = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var compiler_1 = require("@angular/compiler");
    var fatal_linker_error_1 = require("@angular/compiler-cli/linker/src/fatal_linker_error");
    var util_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/util");
    /**
     * A `PartialLinker` that is designed to process `ɵɵngDeclareDirective()` call expressions.
     */
    var PartialDirectiveLinkerVersion1 = /** @class */ (function () {
        function PartialDirectiveLinkerVersion1(sourceUrl, code) {
            this.sourceUrl = sourceUrl;
            this.code = code;
        }
        PartialDirectiveLinkerVersion1.prototype.linkPartialDeclaration = function (constantPool, metaObj) {
            var meta = toR3DirectiveMeta(metaObj, this.code, this.sourceUrl);
            var def = compiler_1.compileDirectiveFromMetadata(meta, constantPool, compiler_1.makeBindingParser());
            return def.expression;
        };
        return PartialDirectiveLinkerVersion1;
    }());
    exports.PartialDirectiveLinkerVersion1 = PartialDirectiveLinkerVersion1;
    /**
     * Derives the `R3DirectiveMetadata` structure from the AST object.
     */
    function toR3DirectiveMeta(metaObj, code, sourceUrl) {
        var typeExpr = metaObj.getValue('type');
        var typeName = typeExpr.getSymbolName();
        if (typeName === null) {
            throw new fatal_linker_error_1.FatalLinkerError(typeExpr.expression, 'Unsupported type, its name could not be determined');
        }
        return {
            typeSourceSpan: createSourceSpan(typeExpr.getRange(), code, sourceUrl),
            type: util_1.wrapReference(typeExpr.getOpaque()),
            typeArgumentCount: 0,
            internalType: metaObj.getOpaque('type'),
            deps: null,
            host: toHostMetadata(metaObj),
            inputs: metaObj.has('inputs') ? metaObj.getObject('inputs').toLiteral(toInputMapping) : {},
            outputs: metaObj.has('outputs') ?
                metaObj.getObject('outputs').toLiteral(function (value) { return value.getString(); }) :
                {},
            queries: metaObj.has('queries') ?
                metaObj.getArray('queries').map(function (entry) { return toQueryMetadata(entry.getObject()); }) :
                [],
            viewQueries: metaObj.has('viewQueries') ?
                metaObj.getArray('viewQueries').map(function (entry) { return toQueryMetadata(entry.getObject()); }) :
                [],
            providers: metaObj.has('providers') ? metaObj.getOpaque('providers') : null,
            fullInheritance: false,
            selector: metaObj.has('selector') ? metaObj.getString('selector') : null,
            exportAs: metaObj.has('exportAs') ?
                metaObj.getArray('exportAs').map(function (entry) { return entry.getString(); }) :
                null,
            lifecycle: {
                usesOnChanges: metaObj.has('usesOnChanges') ? metaObj.getBoolean('usesOnChanges') : false,
            },
            name: typeName,
            usesInheritance: metaObj.has('usesInheritance') ? metaObj.getBoolean('usesInheritance') : false,
        };
    }
    exports.toR3DirectiveMeta = toR3DirectiveMeta;
    /**
     * Decodes the AST value for a single input to its representation as used in the metadata.
     */
    function toInputMapping(value) {
        if (value.isString()) {
            return value.getString();
        }
        var values = value.getArray().map(function (innerValue) { return innerValue.getString(); });
        if (values.length !== 2) {
            throw new fatal_linker_error_1.FatalLinkerError(value.expression, 'Unsupported input, expected a string or an array containing exactly two strings');
        }
        return values;
    }
    /**
     * Extracts the host metadata configuration from the AST metadata object.
     */
    function toHostMetadata(metaObj) {
        if (!metaObj.has('host')) {
            return {
                attributes: {},
                listeners: {},
                properties: {},
                specialAttributes: {},
            };
        }
        var host = metaObj.getObject('host');
        var specialAttributes = {};
        if (host.has('styleAttribute')) {
            specialAttributes.styleAttr = host.getString('styleAttribute');
        }
        if (host.has('classAttribute')) {
            specialAttributes.classAttr = host.getString('classAttribute');
        }
        return {
            attributes: host.has('attributes') ?
                host.getObject('attributes').toLiteral(function (value) { return value.getOpaque(); }) :
                {},
            listeners: host.has('listeners') ?
                host.getObject('listeners').toLiteral(function (value) { return value.getString(); }) :
                {},
            properties: host.has('properties') ?
                host.getObject('properties').toLiteral(function (value) { return value.getString(); }) :
                {},
            specialAttributes: specialAttributes,
        };
    }
    /**
     * Extracts the metadata for a single query from an AST object.
     */
    function toQueryMetadata(obj) {
        var predicate;
        var predicateExpr = obj.getValue('predicate');
        if (predicateExpr.isArray()) {
            predicate = predicateExpr.getArray().map(function (entry) { return entry.getString(); });
        }
        else {
            predicate = predicateExpr.getOpaque();
        }
        return {
            propertyName: obj.getString('propertyName'),
            first: obj.has('first') ? obj.getBoolean('first') : false,
            predicate: predicate,
            descendants: obj.has('descendants') ? obj.getBoolean('descendants') : false,
            emitDistinctChangesOnly: obj.has('emitDistinctChangesOnly') ? obj.getBoolean('emitDistinctChangesOnly') : true,
            read: obj.has('read') ? obj.getOpaque('read') : null,
            static: obj.has('static') ? obj.getBoolean('static') : false,
        };
    }
    function createSourceSpan(range, code, sourceUrl) {
        var sourceFile = new compiler_1.ParseSourceFile(code, sourceUrl);
        var startLocation = new compiler_1.ParseLocation(sourceFile, range.startPos, range.startLine, range.startCol);
        return new compiler_1.ParseSourceSpan(startLocation, startLocation.moveBy(range.endPos - range.startPos));
    }
    exports.createSourceSpan = createSourceSpan;
});
