// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_pipe_linker_1.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_pipe_linker_1", ["require", "exports", "@angular/compiler", "@angular/compiler-cli/linker/src/fatal_linker_error", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.toR3PipeMeta = exports.PartialPipeLinkerVersion1 = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var compiler_1 = require("@angular/compiler");
    var fatal_linker_error_1 = require("@angular/compiler-cli/linker/src/fatal_linker_error");
    var util_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/util");
    /**
     * A `PartialLinker` that is designed to process `ɵɵngDeclarePipe()` call expressions.
     */
    var PartialPipeLinkerVersion1 = /** @class */ (function () {
        function PartialPipeLinkerVersion1() {
        }
        PartialPipeLinkerVersion1.prototype.linkPartialDeclaration = function (constantPool, metaObj) {
            var meta = toR3PipeMeta(metaObj);
            var def = compiler_1.compilePipeFromMetadata(meta);
            return def.expression;
        };
        return PartialPipeLinkerVersion1;
    }());
    exports.PartialPipeLinkerVersion1 = PartialPipeLinkerVersion1;
    /**
     * Derives the `R3PipeMetadata` structure from the AST object.
     */
    function toR3PipeMeta(metaObj) {
        var typeExpr = metaObj.getValue('type');
        var typeName = typeExpr.getSymbolName();
        if (typeName === null) {
            throw new fatal_linker_error_1.FatalLinkerError(typeExpr.expression, 'Unsupported type, its name could not be determined');
        }
        var pure = metaObj.has('pure') ? metaObj.getBoolean('pure') : true;
        return {
            name: typeName,
            type: util_1.wrapReference(typeExpr.getOpaque()),
            internalType: metaObj.getOpaque('type'),
            typeArgumentCount: 0,
            deps: null,
            pipeName: metaObj.getString('name'),
            pure: pure,
        };
    }
    exports.toR3PipeMeta = toR3PipeMeta;
});
