// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injector_linker_1.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injector_linker_1", ["require", "exports", "@angular/compiler", "@angular/compiler-cli/linker/src/fatal_linker_error", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.toR3InjectorMeta = exports.PartialInjectorLinkerVersion1 = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var compiler_1 = require("@angular/compiler");
    var fatal_linker_error_1 = require("@angular/compiler-cli/linker/src/fatal_linker_error");
    var util_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/util");
    /**
     * A `PartialLinker` that is designed to process `ɵɵngDeclareInjector()` call expressions.
     */
    var PartialInjectorLinkerVersion1 = /** @class */ (function () {
        function PartialInjectorLinkerVersion1() {
        }
        PartialInjectorLinkerVersion1.prototype.linkPartialDeclaration = function (constantPool, metaObj) {
            var meta = toR3InjectorMeta(metaObj);
            var def = compiler_1.compileInjector(meta);
            return def.expression;
        };
        return PartialInjectorLinkerVersion1;
    }());
    exports.PartialInjectorLinkerVersion1 = PartialInjectorLinkerVersion1;
    /**
     * Derives the `R3InjectorMetadata` structure from the AST object.
     */
    function toR3InjectorMeta(metaObj) {
        var typeExpr = metaObj.getValue('type');
        var typeName = typeExpr.getSymbolName();
        if (typeName === null) {
            throw new fatal_linker_error_1.FatalLinkerError(typeExpr.expression, 'Unsupported type, its name could not be determined');
        }
        return {
            name: typeName,
            type: util_1.wrapReference(typeExpr.getOpaque()),
            internalType: metaObj.getOpaque('type'),
            providers: metaObj.has('providers') ? metaObj.getOpaque('providers') : null,
            imports: metaObj.has('imports') ? metaObj.getArray('imports').map(function (i) { return i.getOpaque(); }) : [],
        };
    }
    exports.toR3InjectorMeta = toR3InjectorMeta;
});
