// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injectable_linker_1.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injectable_linker_1", ["require", "exports", "@angular/compiler", "@angular/compiler/src/output/output_ast", "@angular/compiler-cli/linker/src/fatal_linker_error", "@angular/compiler-cli/linker/src/file_linker/partial_linkers/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.toR3InjectableMeta = exports.PartialInjectableLinkerVersion1 = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var compiler_1 = require("@angular/compiler");
    var o = require("@angular/compiler/src/output/output_ast");
    var fatal_linker_error_1 = require("@angular/compiler-cli/linker/src/fatal_linker_error");
    var util_1 = require("@angular/compiler-cli/linker/src/file_linker/partial_linkers/util");
    /**
     * A `PartialLinker` that is designed to process `ɵɵngDeclareInjectable()` call expressions.
     */
    var PartialInjectableLinkerVersion1 = /** @class */ (function () {
        function PartialInjectableLinkerVersion1() {
        }
        PartialInjectableLinkerVersion1.prototype.linkPartialDeclaration = function (constantPool, metaObj) {
            var meta = toR3InjectableMeta(metaObj);
            var def = compiler_1.compileInjectable(meta, /* resolveForwardRefs */ false);
            return def.expression;
        };
        return PartialInjectableLinkerVersion1;
    }());
    exports.PartialInjectableLinkerVersion1 = PartialInjectableLinkerVersion1;
    /**
     * Derives the `R3InjectableMetadata` structure from the AST object.
     */
    function toR3InjectableMeta(metaObj) {
        var typeExpr = metaObj.getValue('type');
        var typeName = typeExpr.getSymbolName();
        if (typeName === null) {
            throw new fatal_linker_error_1.FatalLinkerError(typeExpr.expression, 'Unsupported type, its name could not be determined');
        }
        var meta = {
            name: typeName,
            type: util_1.wrapReference(typeExpr.getOpaque()),
            internalType: typeExpr.getOpaque(),
            typeArgumentCount: 0,
            providedIn: metaObj.has('providedIn') ? util_1.extractForwardRef(metaObj.getValue('providedIn')) :
                compiler_1.createR3ProviderExpression(o.literal(null), false),
        };
        if (metaObj.has('useClass')) {
            meta.useClass = util_1.extractForwardRef(metaObj.getValue('useClass'));
        }
        if (metaObj.has('useFactory')) {
            meta.useFactory = metaObj.getOpaque('useFactory');
        }
        if (metaObj.has('useExisting')) {
            meta.useExisting = util_1.extractForwardRef(metaObj.getValue('useExisting'));
        }
        if (metaObj.has('useValue')) {
            meta.useValue = util_1.extractForwardRef(metaObj.getValue('useValue'));
        }
        if (metaObj.has('deps')) {
            meta.deps = metaObj.getArray('deps').map(function (dep) { return util_1.getDependency(dep.getObject()); });
        }
        return meta;
    }
    exports.toR3InjectableMeta = toR3InjectableMeta;
});
