// @angular/compiler-cli/linker/src/file_linker/partial_linkers/util.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/file_linker/partial_linkers/util", ["require", "exports", "@angular/compiler", "@angular/compiler/src/output/output_ast", "@angular/compiler-cli/linker/src/fatal_linker_error"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.extractForwardRef = exports.getDependency = exports.parseEnum = exports.wrapReference = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var compiler_1 = require("@angular/compiler");
    var o = require("@angular/compiler/src/output/output_ast");
    var fatal_linker_error_1 = require("@angular/compiler-cli/linker/src/fatal_linker_error");
    function wrapReference(wrapped) {
        return { value: wrapped, type: wrapped };
    }
    exports.wrapReference = wrapReference;
    /**
     * Parses the value of an enum from the AST value's symbol name.
     */
    function parseEnum(value, Enum) {
        var symbolName = value.getSymbolName();
        if (symbolName === null) {
            throw new fatal_linker_error_1.FatalLinkerError(value.expression, 'Expected value to have a symbol name');
        }
        var enumValue = Enum[symbolName];
        if (enumValue === undefined) {
            throw new fatal_linker_error_1.FatalLinkerError(value.expression, "Unsupported enum value for " + Enum);
        }
        return enumValue;
    }
    exports.parseEnum = parseEnum;
    /**
     * Parse a dependency structure from an AST object.
     */
    function getDependency(depObj) {
        var isAttribute = depObj.has('attribute') && depObj.getBoolean('attribute');
        var token = depObj.getOpaque('token');
        // Normally `attribute` is a string literal and so its `attributeNameType` is the same string
        // literal. If the `attribute` is some other expression, the `attributeNameType` would be the
        // `unknown` type. It is not possible to generate this when linking, since it only deals with JS
        // and not typings. When linking the existence of the `attributeNameType` only acts as a marker to
        // change the injection instruction that is generated, so we just pass the literal string
        // `"unknown"`.
        var attributeNameType = isAttribute ? o.literal('unknown') : null;
        return {
            token: token,
            attributeNameType: attributeNameType,
            host: depObj.has('host') && depObj.getBoolean('host'),
            optional: depObj.has('optional') && depObj.getBoolean('optional'),
            self: depObj.has('self') && depObj.getBoolean('self'),
            skipSelf: depObj.has('skipSelf') && depObj.getBoolean('skipSelf'),
        };
    }
    exports.getDependency = getDependency;
    /**
     * Return an `R3ProviderExpression` that represents either the extracted type reference expression
     * from a `forwardRef` function call, or the type itself.
     *
     * For example, the expression `forwardRef(function() { return FooDir; })` returns `FooDir`. Note
     * that this expression is required to be wrapped in a closure, as otherwise the forward reference
     * would be resolved before initialization.
     *
     * If there is no forwardRef call expression then we just return the opaque type.
     */
    function extractForwardRef(expr) {
        if (!expr.isCallExpression()) {
            return compiler_1.createR3ProviderExpression(expr.getOpaque(), /* isForwardRef */ false);
        }
        var callee = expr.getCallee();
        if (callee.getSymbolName() !== 'forwardRef') {
            throw new fatal_linker_error_1.FatalLinkerError(callee.expression, 'Unsupported expression, expected a `forwardRef()` call or a type reference');
        }
        var args = expr.getArguments();
        if (args.length !== 1) {
            throw new fatal_linker_error_1.FatalLinkerError(expr, 'Unsupported `forwardRef(fn)` call, expected a single argument');
        }
        var wrapperFn = args[0];
        if (!wrapperFn.isFunction()) {
            throw new fatal_linker_error_1.FatalLinkerError(wrapperFn, 'Unsupported `forwardRef(fn)` call, expected its argument to be a function');
        }
        return compiler_1.createR3ProviderExpression(wrapperFn.getFunctionReturnValue().getOpaque(), true);
    }
    exports.extractForwardRef = extractForwardRef;
});
