// @angular/compiler-cli/linker/src/ast/ast_value.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/linker/src/ast/ast_value", ["require", "exports", "tslib", "@angular/compiler", "@angular/compiler-cli/linker/src/fatal_linker_error"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AstValue = exports.AstObject = void 0;
    var tslib_1 = require("tslib");
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var o = require("@angular/compiler");
    var fatal_linker_error_1 = require("@angular/compiler-cli/linker/src/fatal_linker_error");
    /**
     * This helper class wraps an object expression along with an `AstHost` object, exposing helper
     * methods that make it easier to extract the properties of the object.
     *
     * The generic `T` is used as reference type of the expected structure that is represented by this
     * object. It does not achieve full type-safety for the provided operations in correspondence with
     * `T`; its main goal is to provide references to a documented type and ensure that the properties
     * that are read from the object are present.
     *
     * Unfortunately, the generic types are unable to prevent reading an optional property from the
     * object without first having called `has` to ensure that the property exists. This is one example
     * of where full type-safety is not achieved.
     */
    var AstObject = /** @class */ (function () {
        function AstObject(expression, obj, host) {
            this.expression = expression;
            this.obj = obj;
            this.host = host;
        }
        /**
         * Create a new `AstObject` from the given `expression` and `host`.
         */
        AstObject.parse = function (expression, host) {
            var obj = host.parseObjectLiteral(expression);
            return new AstObject(expression, obj, host);
        };
        /**
         * Returns true if the object has a property called `propertyName`.
         */
        AstObject.prototype.has = function (propertyName) {
            return this.obj.has(propertyName);
        };
        /**
         * Returns the number value of the property called `propertyName`.
         *
         * Throws an error if there is no such property or the property is not a number.
         */
        AstObject.prototype.getNumber = function (propertyName) {
            return this.host.parseNumericLiteral(this.getRequiredProperty(propertyName));
        };
        /**
         * Returns the string value of the property called `propertyName`.
         *
         * Throws an error if there is no such property or the property is not a string.
         */
        AstObject.prototype.getString = function (propertyName) {
            return this.host.parseStringLiteral(this.getRequiredProperty(propertyName));
        };
        /**
         * Returns the boolean value of the property called `propertyName`.
         *
         * Throws an error if there is no such property or the property is not a boolean.
         */
        AstObject.prototype.getBoolean = function (propertyName) {
            return this.host.parseBooleanLiteral(this.getRequiredProperty(propertyName));
        };
        /**
         * Returns the nested `AstObject` parsed from the property called `propertyName`.
         *
         * Throws an error if there is no such property or the property is not an object.
         */
        AstObject.prototype.getObject = function (propertyName) {
            var expr = this.getRequiredProperty(propertyName);
            var obj = this.host.parseObjectLiteral(expr);
            return new AstObject(expr, obj, this.host);
        };
        /**
         * Returns an array of `AstValue` objects parsed from the property called `propertyName`.
         *
         * Throws an error if there is no such property or the property is not an array.
         */
        AstObject.prototype.getArray = function (propertyName) {
            var _this = this;
            var arr = this.host.parseArrayLiteral(this.getRequiredProperty(propertyName));
            return arr.map(function (entry) { return new AstValue(entry, _this.host); });
        };
        /**
         * Returns a `WrappedNodeExpr` object that wraps the expression at the property called
         * `propertyName`.
         *
         * Throws an error if there is no such property.
         */
        AstObject.prototype.getOpaque = function (propertyName) {
            return new o.WrappedNodeExpr(this.getRequiredProperty(propertyName));
        };
        /**
         * Returns the raw `TExpression` value of the property called `propertyName`.
         *
         * Throws an error if there is no such property.
         */
        AstObject.prototype.getNode = function (propertyName) {
            return this.getRequiredProperty(propertyName);
        };
        /**
         * Returns an `AstValue` that wraps the value of the property called `propertyName`.
         *
         * Throws an error if there is no such property.
         */
        AstObject.prototype.getValue = function (propertyName) {
            return new AstValue(this.getRequiredProperty(propertyName), this.host);
        };
        /**
         * Converts the AstObject to a raw JavaScript object, mapping each property value (as an
         * `AstValue`) to the generic type (`T`) via the `mapper` function.
         */
        AstObject.prototype.toLiteral = function (mapper) {
            var e_1, _a;
            var result = {};
            try {
                for (var _b = tslib_1.__values(this.obj), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var _d = tslib_1.__read(_c.value, 2), key = _d[0], expression = _d[1];
                    result[key] = mapper(new AstValue(expression, this.host));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return result;
        };
        /**
         * Converts the AstObject to a JavaScript Map, mapping each property value (as an
         * `AstValue`) to the generic type (`T`) via the `mapper` function.
         */
        AstObject.prototype.toMap = function (mapper) {
            var e_2, _a;
            var result = new Map();
            try {
                for (var _b = tslib_1.__values(this.obj), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var _d = tslib_1.__read(_c.value, 2), key = _d[0], expression = _d[1];
                    result.set(key, mapper(new AstValue(expression, this.host)));
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return result;
        };
        AstObject.prototype.getRequiredProperty = function (propertyName) {
            if (!this.obj.has(propertyName)) {
                throw new fatal_linker_error_1.FatalLinkerError(this.expression, "Expected property '" + propertyName + "' to be present.");
            }
            return this.obj.get(propertyName);
        };
        return AstObject;
    }());
    exports.AstObject = AstObject;
    /**
     * This helper class wraps an `expression`, exposing methods that use the `host` to give
     * access to the underlying value of the wrapped expression.
     *
     * The generic `T` is used as reference type of the expected type that is represented by this value.
     * It does not achieve full type-safety for the provided operations in correspondence with `T`; its
     * main goal is to provide references to a documented type.
     */
    var AstValue = /** @class */ (function () {
        function AstValue(expression, host) {
            this.expression = expression;
            this.host = host;
        }
        /**
         * Get the name of the symbol represented by the given expression node, or `null` if it is not a
         * symbol.
         */
        AstValue.prototype.getSymbolName = function () {
            return this.host.getSymbolName(this.expression);
        };
        /**
         * Is this value a number?
         */
        AstValue.prototype.isNumber = function () {
            return this.host.isNumericLiteral(this.expression);
        };
        /**
         * Parse the number from this value, or error if it is not a number.
         */
        AstValue.prototype.getNumber = function () {
            return this.host.parseNumericLiteral(this.expression);
        };
        /**
         * Is this value a string?
         */
        AstValue.prototype.isString = function () {
            return this.host.isStringLiteral(this.expression);
        };
        /**
         * Parse the string from this value, or error if it is not a string.
         */
        AstValue.prototype.getString = function () {
            return this.host.parseStringLiteral(this.expression);
        };
        /**
         * Is this value a boolean?
         */
        AstValue.prototype.isBoolean = function () {
            return this.host.isBooleanLiteral(this.expression);
        };
        /**
         * Parse the boolean from this value, or error if it is not a boolean.
         */
        AstValue.prototype.getBoolean = function () {
            return this.host.parseBooleanLiteral(this.expression);
        };
        /**
         * Is this value an object literal?
         */
        AstValue.prototype.isObject = function () {
            return this.host.isObjectLiteral(this.expression);
        };
        /**
         * Parse this value into an `AstObject`, or error if it is not an object literal.
         */
        AstValue.prototype.getObject = function () {
            return AstObject.parse(this.expression, this.host);
        };
        /**
         * Is this value an array literal?
         */
        AstValue.prototype.isArray = function () {
            return this.host.isArrayLiteral(this.expression);
        };
        /**
         * Parse this value into an array of `AstValue` objects, or error if it is not an array literal.
         */
        AstValue.prototype.getArray = function () {
            var _this = this;
            var arr = this.host.parseArrayLiteral(this.expression);
            return arr.map(function (entry) { return new AstValue(entry, _this.host); });
        };
        /**
         * Is this value a function expression?
         */
        AstValue.prototype.isFunction = function () {
            return this.host.isFunctionExpression(this.expression);
        };
        /**
         * Extract the return value as an `AstValue` from this value as a function expression, or error if
         * it is not a function expression.
         */
        AstValue.prototype.getFunctionReturnValue = function () {
            return new AstValue(this.host.parseReturnValue(this.expression), this.host);
        };
        AstValue.prototype.isCallExpression = function () {
            return this.host.isCallExpression(this.expression);
        };
        AstValue.prototype.getCallee = function () {
            return new AstValue(this.host.parseCallee(this.expression), this.host);
        };
        AstValue.prototype.getArguments = function () {
            var _this = this;
            var args = this.host.parseArguments(this.expression);
            return args.map(function (arg) { return new AstValue(arg, _this.host); });
        };
        /**
         * Return the `TExpression` of this value wrapped in a `WrappedNodeExpr`.
         */
        AstValue.prototype.getOpaque = function () {
            return new o.WrappedNodeExpr(this.expression);
        };
        /**
         * Get the range of the location of this value in the original source.
         */
        AstValue.prototype.getRange = function () {
            return this.host.getRange(this.expression);
        };
        return AstValue;
    }());
    exports.AstValue = AstValue;
});
