// @angular/cdk/schematics/ng-update/devkit-file-system.mjs
"use strict";
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevkitFileSystem = void 0;
const core_1 = require("@angular-devkit/core");
const file_system_1 = require("../update-tool/file-system");
const path = require("path");
/**
 * File system that leverages the virtual tree from the CLI devkit. This file
 * system is commonly used by `ng update` migrations that run as part of the
 * Angular CLI.
 */
class DevkitFileSystem extends file_system_1.FileSystem {
    constructor(_tree) {
        super();
        this._tree = _tree;
        this._updateRecorderCache = new Map();
    }
    resolve(...segments) {
        // Note: We use `posix.resolve` as the devkit paths are using posix separators.
        return core_1.normalize(path.posix.resolve('/', ...segments.map(core_1.normalize)));
    }
    edit(filePath) {
        if (this._updateRecorderCache.has(filePath)) {
            return this._updateRecorderCache.get(filePath);
        }
        const recorder = this._tree.beginUpdate(filePath);
        this._updateRecorderCache.set(filePath, recorder);
        return recorder;
    }
    commitEdits() {
        this._updateRecorderCache.forEach(r => this._tree.commitUpdate(r));
        this._updateRecorderCache.clear();
    }
    exists(fileOrDirPath) {
        // We need to check for both file or directory existence, in order
        // to comply with the expectation from the TypeScript compiler.
        return this._tree.exists(fileOrDirPath) || this._isExistingDirectory(fileOrDirPath);
    }
    overwrite(filePath, content) {
        this._tree.overwrite(filePath, content);
    }
    create(filePath, content) {
        this._tree.create(filePath, content);
    }
    delete(filePath) {
        this._tree.delete(filePath);
    }
    read(filePath) {
        const buffer = this._tree.read(filePath);
        return buffer !== null ? buffer.toString() : null;
    }
    readDirectory(dirPath) {
        const { subdirs: directories, subfiles: files } = this._tree.getDir(dirPath);
        return { directories, files };
    }
    _isExistingDirectory(dirPath) {
        if (dirPath === core_1.NormalizedRoot) {
            return true;
        }
        const parent = core_1.dirname(dirPath);
        const dirName = core_1.basename(dirPath);
        // TypeScript also checks potential entry points, so e.g. importing
        // package.json will result in a lookup of /package.json/package.json
        // and /package.json/index.ts. In order to avoid failure, we check if
        // the parent is an existing file and return false, if that is the case.
        if (this._tree.exists(parent)) {
            return false;
        }
        const dir = this._tree.getDir(parent);
        return dir.subdirs.indexOf(dirName) !== -1;
    }
}
exports.DevkitFileSystem = DevkitFileSystem;
