// @angular/compiler-cli/src/ngtsc/metadata/src/resource_registry.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/metadata/src/resource_registry", ["require", "exports", "tslib"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ResourceRegistry = exports.isExternalResource = void 0;
    var tslib_1 = require("tslib");
    function isExternalResource(resource) {
        return resource.path !== null;
    }
    exports.isExternalResource = isExternalResource;
    /**
     * Tracks the mapping between external template/style files and the component(s) which use them.
     *
     * This information is produced during analysis of the program and is used mainly to support
     * external tooling, for which such a mapping is challenging to determine without compiler
     * assistance.
     */
    var ResourceRegistry = /** @class */ (function () {
        function ResourceRegistry() {
            this.externalTemplateToComponentsMap = new Map();
            this.componentToTemplateMap = new Map();
            this.componentToStylesMap = new Map();
            this.externalStyleToComponentsMap = new Map();
        }
        ResourceRegistry.prototype.getComponentsWithTemplate = function (template) {
            if (!this.externalTemplateToComponentsMap.has(template)) {
                return new Set();
            }
            return this.externalTemplateToComponentsMap.get(template);
        };
        ResourceRegistry.prototype.registerResources = function (resources, component) {
            var e_1, _a;
            if (resources.template !== null) {
                this.registerTemplate(resources.template, component);
            }
            try {
                for (var _b = tslib_1.__values(resources.styles), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var style = _c.value;
                    this.registerStyle(style, component);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        };
        ResourceRegistry.prototype.registerTemplate = function (templateResource, component) {
            var path = templateResource.path;
            if (path !== null) {
                if (!this.externalTemplateToComponentsMap.has(path)) {
                    this.externalTemplateToComponentsMap.set(path, new Set());
                }
                this.externalTemplateToComponentsMap.get(path).add(component);
            }
            this.componentToTemplateMap.set(component, templateResource);
        };
        ResourceRegistry.prototype.getTemplate = function (component) {
            if (!this.componentToTemplateMap.has(component)) {
                return null;
            }
            return this.componentToTemplateMap.get(component);
        };
        ResourceRegistry.prototype.registerStyle = function (styleResource, component) {
            var path = styleResource.path;
            if (!this.componentToStylesMap.has(component)) {
                this.componentToStylesMap.set(component, new Set());
            }
            if (path !== null) {
                if (!this.externalStyleToComponentsMap.has(path)) {
                    this.externalStyleToComponentsMap.set(path, new Set());
                }
                this.externalStyleToComponentsMap.get(path).add(component);
            }
            this.componentToStylesMap.get(component).add(styleResource);
        };
        ResourceRegistry.prototype.getStyles = function (component) {
            if (!this.componentToStylesMap.has(component)) {
                return new Set();
            }
            return this.componentToStylesMap.get(component);
        };
        ResourceRegistry.prototype.getComponentsWithStyle = function (styleUrl) {
            if (!this.externalStyleToComponentsMap.has(styleUrl)) {
                return new Set();
            }
            return this.externalStyleToComponentsMap.get(styleUrl);
        };
        return ResourceRegistry;
    }());
    exports.ResourceRegistry = ResourceRegistry;
});
